// news_cards/index.js
const STATE = require('STATE')
const statedb = STATE(__filename)

const { get } = statedb(fallback_module)

module.exports = news_cards

async function news_cards (opts, protocol) {
  const { sid } = opts
  await get(sid)

  const el = document.createElement('div')
  let current_path = ''
  let send

  function handle_message (msg) {
    if (msg.type === 'provision') render(msg.data)
  }

  if (protocol) send = protocol(handle_message)

  el.addEventListener('click', handle_card_click)

  return el

  function handle_card_click () {
    if (send && current_path) {
      send({ type: 'card_clicked', data: { path: current_path } })
    }
  }

  function render (card_data) {
    const data = card_data.data || {}
    current_path = card_data.path || ''
    const date = data.date || ''
    const title = data.title || 'Untitled Story'
    const author = data.author || 'Anonymous'
    const description = data.description || ''
    const tags = Array.isArray(data.tags) ? data.tags : []
    const tags_html = tags.map(render_tag).join('')

    el.innerHTML = `
      <div class="news-card">
        <div class="news-avatar">${author[0] || 'A'}</div>
        <div class="news-content">
          <div class="news-meta-top">
            <span class="news-author">${author}</span>
            <span class="news-separator">•</span>
            <span class="news-date-text">${date}</span>
          </div>
          <h3 class="news-title">${title}</h3>
          <p class="news-description">${description}</p>
          <div class="news-meta-bottom">
            <div class="news-tags">${tags_html}</div>
          </div>
        </div>
      </div>
    `
  }

  function render_tag (tag) {
    return `<span class="news-tag-pill">${tag}</span>`
  }
}

function fallback_module () {
  return { api: fallback_instance }

  function fallback_instance () {
    const drive = {
      'runtime/': {
        'card-0.json': { raw: '{}' }
      }
    }
    return { drive }
  }
}
