// news/dummy_content.js
module.exports = {
    '/my-stories/story-1': {
        title: 'My First Post on P2P Stories',
        author: 'You',
        date: '3h ago',
        description: 'Excited to join this decentralized blogging platform! Here I share my thoughts on why peer-to-peer networks are the future...',
        tags: [],
        color: '#a78bfa',
        content: '<p>Excited to join this decentralized blogging platform!</p>'
    },
    '/my-stories/story-2': {
        title: 'Building a P2P Network from Scratch',
        author: 'You',
        date: '2d ago',
        description: 'Learn how to create a decentralized peer-to-peer network using modern web technologies. This guide covers WebRTC, DHTs...',
        tags: ['#Tech', '#Tutorial'],
        color: '#a78bfa',
        content: '<p>Learn how to create a decentralized peer-to-peer network...</p>'
    },
    '/my-stories/story-3': {
        title: 'The Future of Decentralized Publishing',
        author: 'You',
        date: '5d ago',
        description: 'Traditional platforms face centralization issues. Let\'s explore how blockchain + P2P tech give creators freedom...',
        tags: ['#Opinion', '#Blockchain'],
        color: '#a78bfa',
        content: '<p>Traditional platforms face centralization issues...</p>'
    },
    '/my-stories/story-4': {
        title: 'Why Decentralization Matters in 2025',
        author: 'You',
        date: '1w ago',
        description: 'As we enter a hyper-connected world, decentralization becomes essential for privacy, security and independence...',
        tags: ['#Opinion', '#Tech'],
        color: '#a78bfa',
        content: '<p>As we enter a hyper-connected world...</p>'
    },
    '/feeds/off-the-grid/mesh-network': {
        title: 'Building a mesh network in your neighborhood',
        author: 'Off the Grid',
        date: '2h ago',
        description: 'Community-owned internet is possible. Here\'s how we set up a local mesh network connecting 50+ homes without ISPs...',
        tags: ['#tech', '#p2p'],
        color: '#e0e7ff',
        content: '<p>Community-owned internet is possible...</p>'
    },
    '/feeds/off-the-grid/fediverse': {
        title: 'Why I left social media for the fediverse',
        author: 'Off the Grid',
        date: '5h ago',
        description: 'After 10 years on corporate platforms, I finally made the switch. Here\'s what I learned about digital independence...',
        tags: ['#opinion', '#blockchain'],
        color: '#fecaca',
        content: '<p>After 10 years on corporate platforms...</p>'
    },
    '/feeds/off-the-grid/self-hosting': {
        title: 'Self-hosting 101: Start with these apps',
        author: 'Off the Grid',
        date: '1d ago',
        description: 'You don\'t need to be a sysadmin to take control of your data. Here are 5 easy self-hosted apps anyone can set up...',
        tags: ['#tutorials'],
        color: '#22d3ee',
        content: '<p>You don\'t need to be a sysadmin...</p>'
    },
    '/feeds/hackers-digest/code-coffee': {
        title: 'Code & Coffee: Episode 42',
        author: 'Hacker Digest',
        date: '3h ago',
        description: 'This week we discuss why your try-catch blocks are hiding bugs. Stop swallowing errors, log everything, and drink better coffee.',
        tags: ['#podcast', '#debugging', '#coffee'],
        color: '#fcd34d',
        content: `
      <h2>Espresso & Error Handling</h2>
      <p>This week we discuss why your try-catch blocks are hiding bugs.</p>
    `
    },
    '/feeds/hackers-digest/system-design': {
        title: 'System Design: Scaling to 1M Users',
        author: 'Hacker Digest',
        date: '6h ago',
        description: 'A deep dive into sharding databases and load balancing. Horizontal vs Vertical Scaling explained for modern distributed systems.',
        tags: ['#system-design', '#scaling', '#architecture'],
        color: '#6ee7b7',
        content: `
      <h2>Horizontal vs Vertical Scaling</h2>
      <p>A deep dive into sharding databases and load balancing.</p>
    `
    },
    '/feeds/peer-review/network-notes': {
        title: 'Network Topologies in 2025',
        author: 'Peer Review',
        date: '12h ago',
        description: 'Mesh networks are finally becoming viable for consumer apps. We explore the latest advancements in DHTs and gossip protocols.',
        tags: ['#networking', '#mesh', '#future'],
        color: '#93c5fd',
        content: `
      <p>Mesh networks are finally becoming viable for consumer apps.</p>
    `
    },
    '/feeds/peer-review/security-chronicles': {
        title: 'Security Chronicles',
        author: 'Peer Review',
        date: '1d ago',
        description: 'A collection of the latest security findings, vulnerabilities, and privacy discussions from around the decentralized web.',
        tags: ['#security', '#privacy', '#collection'],
        color: '#fca5a5',
        content: ''
    },
    '/feeds/peer-review/security-chronicles/privacy-matters': {
        title: 'Why Privacy Matters',
        author: 'Security Chronicles',
        date: '2d ago',
        description: 'Privacy is not about hiding; it is about consent. Understanding the fundamental human right to digital privacy in the age of surveillance.',
        tags: ['#privacy', '#rights', '#philosophy'],
        color: '#fca5a5',
        content: `
      <p>Privacy is not about hiding; it's about consent.</p>
    `
    },
    '/feeds/peer-review/security-chronicles/zero-trust': {
        title: 'Zero Trust Architecture',
        author: 'Security Chronicles',
        date: '3d ago',
        description: 'Traditional perimeter security is obsolete. Implement zero trust security in your organization by verifying every request, regardless of origin.',
        tags: ['#security', '#cloud', '#enterprise'],
        color: '#fca5a5',
        content: `
      <p>Traditional perimeter security is obsolete.</p>
    `
    },
    '/lists/best-of-tech': {
        title: 'Best Tech of 2024',
        author: 'TechDaily',
        date: '2024-12-30',
        content: `
      <p>1. P2P News App</p>
      <p>2. The rest...</p>
    `
    },
    '/lists/morning-read': {
        title: 'Morning Read: Stoicism for Devs',
        author: 'Daily Stoic',
        date: '2025-01-01',
        content: `
      <p>Accept the bugs you cannot fix. Have courage to fix the ones you can.</p>
    `
    }
}
