// news/index.js
const STATE = require('STATE')
const statedb = STATE(__filename)

const { get } = statedb(fallback_module)
const menu_sidebar = require('menu_sidebar')
const content_parser = require('newsfeed_view/content_parser')
const graphdb = require('./graphdb')

module.exports = news_app

async function news_app (opts, protocol) {
  const { sid } = opts
  const { id, sdb } = await get(sid)
  const { drive } = sdb

  const by = id
  let db = null
  let msg_id = 0
  let send_to_sidebar = null

  let active_path = ''
  let active_tab = 'news'

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="container">
      <div class="sidebar"></div>
      <div class="main-viewer main"></div>
    </div>
  `

  const sidebar_el = shadow.querySelector('.sidebar')
  const main_viewer = shadow.querySelector('.main-viewer')

  const shell_sheet = new CSSStyleSheet()
  const layout_sheet = new CSSStyleSheet()
  const card_sheet = new CSSStyleSheet()
  const sheet = new CSSStyleSheet()

  shadow.adoptedStyleSheets = [shell_sheet, layout_sheet, card_sheet, sheet]

  const css_files = await Promise.all([
    drive.get('theme/shell.css'),
    drive.get('theme/layout.css'),
    drive.get('theme/news-card.css')
  ])
  const [shell_file, layout_file, card_file] = css_files

  if (shell_file && shell_file.raw) shell_sheet.replaceSync(shell_file.raw)
  if (layout_file && layout_file.raw) layout_sheet.replaceSync(layout_file.raw)
  if (card_file && card_file.raw) card_sheet.replaceSync(card_file.raw)

  const subs = await sdb.watch(onbatch)

  if (subs && subs.length > 0) {
    const menu_el = await menu_sidebar({ sid: subs[0].sid }, sidebar_protocol)
    sidebar_el.appendChild(menu_el)
  }

  render_main_view()

  shadow.addEventListener('click', handle_shadow_click)
  shadow.addEventListener('submit', handle_shadow_submit)
  shadow.addEventListener('input', handle_shadow_input)

  return el

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(fetch_path_raw_data))
      const valid_data = data.filter(filter_valid_data)
      if (valid_data.length > 0) {
        if (type === 'theme') inject(valid_data)
        if (type === 'entries') on_entries(valid_data)
      }
    }

    function fetch_path_raw_data (path) {
      return drive.get(path).then(extract_raw_if_exists)
    }

    function extract_raw_if_exists (file) {
      return file ? file.raw : null
    }

    function filter_valid_data (d) {
      return d !== null
    }
  }

  function inject (data) {
    if (Array.isArray(data)) {
      sheet.replaceSync(data.join('\n'))
    }
  }

  function sidebar_protocol (send) {
    send_to_sidebar = send
    return on_sidebar_message
  }

  function on_sidebar_message (msg) {
    const { type, data } = msg
    if (type === 'selection_changed') {
      const { selected } = data
      if (selected && selected.length > 0) {
        const path = normalize_path(selected[0])
        if (path !== active_path) {
          active_path = path
          const node_type = db ? db.type(path) : null
          if (node_type === 'subscribed-blog') active_tab = 'blog'
          else active_tab = 'news'
          render_content(path)
        }
      }
    }
    if (type.startsWith('db_')) {
      handle_db_request(msg, send_to_sidebar)
    }
  }

  function handle_shadow_click (e) {
    if (e.target.closest('.news-fab')) {
      render_write_page_view(active_path)
    }
    if (e.target.classList.contains('tab')) {
      handle_tab_click(e.target.dataset.tab)
    }
    const card_target = e.target.closest('.news-card')
    if (card_target) {
      handle_card_click(card_target.dataset.path)
    }
    if (e.target.closest('.back-btn')) {
      handle_back_click()
    }
  }

  function handle_tab_click (tab) {
    if (tab && tab !== active_tab) {
      active_tab = tab
      render_content(active_path)
    }
  }

  function handle_card_click (path) {
    if (path) {
      render_article_view(path)
    }
  }

  function handle_back_click () {
    render_content(active_path)
  }

  function handle_shadow_submit (e) {
    if (e.target.id === 'write-story-form') {
      e.preventDefault()
      const create_story_data = {
        title: e.target.title.value,
        content: e.target.content.value,
        blog: e.target.blog.value
      }
      handle_publish(create_story_data)
    }
  }

  function handle_shadow_input (e) {
    if (e.target.classList.contains('input-content')) {
      handle_content_input(e.target.value)
    }
  }

  function handle_content_input (val) {
    const words = val.trim() === '' ? 0 : val.trim().split(/\s+/).length
    const wc = shadow.getElementById('word-count-span')
    const rt = shadow.getElementById('read-time-span')
    if (wc) wc.textContent = `${words} words`
    if (rt) rt.textContent = `~${Math.ceil(words / 200)} min read`
  }

  function render_main_view () {
    main_viewer.innerHTML = `
      <div class="empty-container">
        <h2 class="empty-title">Select a blog to start</h2>
        <p class="empty-hint">← Explore your blogs and subscriptions in the sidebar</p>
      </div>
    `
  }

  function handle_publish (data) {
    const new_story = {
      title: data.title,
      content: data.content,
      author: 'You',
      date: new Date().toLocaleDateString(),
      description: data.content.slice(0, 100) + '...',
      tags: ['#local', '#' + data.blog.replace(/\s+/g, '')],
      color: '#6366f1'
    }
    save_local_story(new_story)
    alert('Story published locally!')
    render_content(active_path)
  }

  async function on_entries (data) {
    if (!data || !data[0]) {
      db = graphdb({})
      notify_db_initialized({})
      return
    }
    let parsed_data = {}
    try {
      if (typeof data[0] === 'string') {
        parsed_data = JSON.parse(data[0])
      } else if (data[0].raw && typeof data[0].raw === 'string') {
        parsed_data = JSON.parse(data[0].raw)
      } else {
        parsed_data = data[0]
      }
    } catch (e) { }
    db = graphdb(parsed_data)
    notify_db_initialized(parsed_data)
  }

  function notify_db_initialized (entries) {
    if (send_to_sidebar) {
      const head = [by, 'menu_sidebar', msg_id++]
      send_to_sidebar({ head, type: 'db_initialized', data: { entries } })
    }
  }

  async function render_content (path) {
    const node_type = db ? db.type(path) : null
    const node = db ? db.get(path) : null

    if (!node) return render_main_view()

    let tabs_html = ''
    if (node_type !== 'discover') {
      tabs_html = `
        <div class="view-tabs">
          <button class="tab ${active_tab === 'news' ? 'active' : ''}" data-tab="news">News</button>
          <button class="tab ${active_tab === 'blog' ? 'active' : ''}" data-tab="blog">Blog</button>
        </div>
      `
    }

    main_viewer.innerHTML = `
      <div class="content-container">
        ${tabs_html}
        <div class="tab-content"></div>
      </div>
    `
    const content_el = main_viewer.querySelector('.tab-content')

    if (active_tab === 'news') {
      await render_news_tab(path, content_el)
    } else {
      await render_blog_tab(path, content_el)
    }
  }

  async function render_news_tab (path, target) {
    const node = db.get(path)
    const node_subs = node.subs || []
    const fetched_items = []

    for (const sub_path of node_subs) {
      const sub_node = db.get(sub_path)
      if (sub_node && sub_node.posts) {
        for (const post_path of sub_node.posts) {
          const post_data = await fetch_post_data(post_path)
          if (post_data) fetched_items.push({ path: post_path, data: post_data })
        }
      }
    }

    if (node.posts) {
      for (const post_path of node.posts) {
        const post_data = await fetch_post_data(post_path)
        if (post_data) fetched_items.push({ path: post_path, data: post_data })
      }
    }

    if (fetched_items.length > 0) {
      target.innerHTML = render_card_list(fetched_items)
    } else {
      target.innerHTML = `<div class="empty-container"><p>No news to show from subscriptions.</p></div>`
    }
  }

  async function render_blog_tab (path, target) {
    const node = db.get(path)
    const node_type = db.type(path)
    let fetched_items = []

    if (node_type === 'my-blog') {
      const local_stories = get_local_stories()
      fetched_items = local_stories.map(map_local_story)
    } else if (node.posts) {
      for (const post_path of node.posts) {
        const post_data = await fetch_post_data(post_path)
        if (post_data) fetched_items.push({ path: post_path, data: post_data })
      }
    }

    if (fetched_items.length > 0 || node_type === 'my-blog') {
      target.innerHTML = render_card_list(fetched_items)
      if (node_type === 'my-blog') {
        target.insertAdjacentHTML('beforeend', '<div class="news-fab">+</div>')
      }
    } else {
      target.innerHTML = `<div class="empty-container"><p>This blog hasn't posted anything yet.</p></div>`
    }
  }

  function render_card_list (items) {
    if (items.length === 0) {
      return '<div class="empty-container"><p>No posts yet.</p></div>'
    }
    const cards = items.map(render_card_html).join('')
    return `<div class="newsfeed-list">${cards}</div>`
  }

  function render_card_html (item) {
    const data = item.data || {}
    const date = data.date || ''
    const title = data.title || 'Untitled Story'
    const author = data.author || 'Anonymous'
    const description = data.description || ''
    const tags = Array.isArray(data.tags) ? data.tags : (typeof data.tags === 'string' ? data.tags.split(/\s+/).filter(Boolean) : [])
    const tags_html = tags.map(render_tag_html).join('')
    
    return `
      <div class="news-card" data-path="${item.path}">
        <div class="news-avatar">${author[0] || 'A'}</div>
        <div class="news-content">
          <div class="news-meta-top">
            <span class="news-author">${author}</span>
            <span class="news-separator">•</span>
            <span class="news-date-text">${date}</span>
          </div>
          <h3 class="news-title">${title}</h3>
          <p class="news-description">${description}</p>
          <div class="news-meta-bottom">
            <div class="news-tags">${tags_html}</div>
          </div>
        </div>
      </div>
    `
  }

  function render_tag_html (tag) {
    return `<span class="news-tag-pill">${tag}</span>`
  }

  async function render_article_view (post_path) {
    const post_data = await fetch_post_data(post_path)
    if (!post_data) {
      main_viewer.innerHTML = '<div class="empty-container"><p>Content not available.</p></div>'
      return
    }
    const content_html = (post_data.content || '')
      .split('\n\n')
      .map(parse_block)
      .join('')

    main_viewer.innerHTML = `
      <div class="content-container article-view">
        <article class="news-container">
          <button class="back-btn">← Back to feed</button>
          <header class="news-header">
            <div>
              <h2 class="news-title article-title-main">${post_data.title || 'Untitled'}</h2>
              <div class="news-subheader">
                <span>By <strong>${post_data.author || 'Anonymous'}</strong></span> • <span>${post_data.date || 'Unknown Date'}</span>
              </div>
            </div>
          </header>
          <div class="article-body">${content_html}</div>
        </article>
      </div>
    `
  }

  function render_write_page_view (path) {
    const node = db ? db.get(path) : null
    const blog_name = node ? node.name : 'Main Blog'
    
    main_viewer.innerHTML = `
      <div class="write-page-container">
        <div class="section-header">
          <h1>Write a Story</h1>
          <p>Create something amazing for <strong>${blog_name}</strong></p>
        </div>
        
        <div class="card space-y-8">
          <form id="write-story-form">
            <div class="input-group">
              <label>Target Blog</label>
              <input type="text" name="blog" value="${blog_name}" readonly class="blog-select readonly-blog-select">
            </div>
            
            <div class="input-group">
              <label>Story Title</label>
              <input type="text" name="title" placeholder="Enter your story title..." class="input-title" required autofocus>
            </div>
            
            <div class="divider"></div>
            
            <div class="input-group">
              <label>Content</label>
              <textarea name="content" placeholder="Start writing your story here..." class="input-content" required></textarea>
            </div>
            
            <div class="word-count">
              <span id="word-count-span">0 words</span>
              <span id="read-time-span">~0 min read</span>
            </div>
            
            <div class="actions">
              <button type="submit" class="btn-publish">Publish Story</button>
              <span class="action-text">Saved to local draft</span>
            </div>
          </form>
        </div>
      </div>
    `
  }

  function map_local_story (s) {
    return { path: `local-${Math.random()}`, data: s }
  }

  async function fetch_post_data (post_path) {
    try {
      const file = await drive.get(post_path)
      if (file && file.raw) {
        const parsed = content_parser({ raw: file.raw })
        return parsed
      }
    } catch (e) { }
    return null
  }

  function normalize_path (path) {
    if (!path) return '/'
    if (typeof path !== 'string') return '/'
    return path.split('|').pop() || '/'
  }

  function get_local_stories () {
    try {
      const stories = localStorage.getItem('p2p_stories')
      return stories ? JSON.parse(stories) : []
    } catch (e) { return [] }
  }

  function save_local_story (story) {
    const stories = get_local_stories()
    stories.unshift(story)
    localStorage.setItem('p2p_stories', JSON.stringify(stories))
  }

  async function handle_db_request (request_msg, send) {
    const { head: request_head, type: operation, data: params } = request_msg
    let result
    if (!db) { send_response(request_head, null); return }

    if (operation === 'db_get') result = db.get(params.path)
    else if (operation === 'db_has') result = db.has(params.path)
    else if (operation === 'db_is_empty') result = db.is_empty()
    else if (operation === 'db_root') result = db.root()
    else if (operation === 'db_keys') result = db.keys()
    else if (operation === 'db_raw') result = db.raw()
    else result = null

    send_response(request_head, result)

    function send_response (request_head, result) {
      const response_head = [by, 'menu_sidebar', msg_id++]
      send({ head: response_head, refs: { cause: request_head }, type: 'db_response', data: { result } })
    }
  }

  function parse_block (block) {
    block = block.trim()
    if (!block) return ''
    if (block.startsWith('# ')) return `<h1>${block.slice(2)}</h1>`
    if (block.startsWith('## ')) return `<h2>${block.slice(3)}</h2>`
    if (block.startsWith('### ')) return `<h3>${block.slice(4)}</h3>`
    if (block.startsWith('- ')) {
      const items = block.split('\n').map(format_list_item).join('')
      return `<ul>${items}</ul>`
    }
    let p = block
    p = p.replace(/\*\*(.*?)\*\*/g, '<strong>$1</strong>')
    p = p.replace(/\*(.*?)\*/g, '<em>$1</em>')
    p = p.replace(/\[(.*?)\]\((.*?)\)/g, '<a href="$2">$1</a>')
    return `<p>${p}</p>`
  }

  function format_list_item (line) {
    return `<li>${line.replace(/^- /, '')}</li>`
  }
}

function fallback_module () {
  const _ = {
    menu_sidebar: { $: '' },
    './graphdb': { $: '' },
    'newsfeed_view/content_parser': { $: '' }
  }

  return { _, api: fallback_instance }

  function fallback_instance (opts) {
    const _ = {
      menu_sidebar: {
        0: '',
        mapping: {
          style: 'style',
          entries: 'entries',
          runtime: 'runtime',
          mode: 'mode',
          flags: 'flags',
          keybinds: 'keybinds',
          undo: 'undo',
          data: 'data'
        }
      },
      './graphdb': { 0: '' },
      'newsfeed_view/content_parser': { 0: '' }
    }

    const drive = {
      'entries/': { 'entries.json': { $ref: 'entries.json' } },
      'data/': {},
      'news_cards/': {
        'news-card.css': { $ref: '../news_cards/news-card.css' }
      },
      'theme/': {
        'shell.css': {
          raw: `
    body { margin: 0; padding: 0; overflow: hidden; }
    .container {
      display: flex;
      width: 100vw;
      height: 100vh;
      overflow: hidden;
      font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif;
    }
    .sidebar {
      width: 250px;
      min-width: 250px;
      border-right: 1px solid #e5e7eb;
      background: #f9fafb;
      display: flex;
      flex-direction: column;
    }
    .main {
      flex: 1;
      overflow: auto;
      padding: 0;
      background: #ffffff;
    }
          `
        },
        'layout.css': {
          $ref: 'layout.css'
        },
        'news-card.css': {
          $ref: '../news_cards/news-card.css'
        }
      },
      'style/': {},
      'runtime/': {
        'viewer_data.json': { raw: '{}' }
      },
      'mode/': {},
      'flags/': {},
      'keybinds/': {},
      'undo/': {},
      'posts/': {
        'tech-thoughts-1': { $ref: 'data/tech-thoughts-1.md' },
        'tech-thoughts-2': { $ref: 'data/tech-thoughts-2.md' },
        'p2p-experiments-1': { $ref: 'data/p2p-experiments-1.md' },
        'p2p-experiments-2': { $ref: 'data/p2p-experiments-2.md' },
        'security-notes-1': { $ref: 'data/security-notes-1.md' },
        'hackers-digest-1': { $ref: 'data/hackers-digest-1.md' },
        'hackers-digest-2': { $ref: 'data/hackers-digest-2.md' },
        'off-the-grid-1': { $ref: 'data/off-the-grid-1.md' },
        'off-the-grid-2': { $ref: 'data/off-the-grid-2.md' },
        'peer-review-1': { $ref: 'data/peer-review-1.md' },
        'cryptography-weekly-1': { $ref: 'data/cryptography-weekly-1.md' },
        'random-blog-a-1': { $ref: 'data/random-blog-a-1.md' },
        'random-blog-b-1': { $ref: 'data/random-blog-b-1.md' },
        'random-blog-c-1': { $ref: 'data/random-blog-c-1.md' }
      }
    }

    return { _, drive }
  }
}
