// news/index.js
const STATE = require('STATE')
const statedb = STATE(__filename)

const { get } = statedb(fallback_module)
const menu_sidebar = require('menu_sidebar')
const newsfeed_view = require('newsfeed_view')
const write_page = require('write_page')
const content_parser = require('newsfeed_view/content_parser')
const graphdb = require('./graphdb')
const net_helper = require('net_helper')

module.exports = news_app

async function news_app (opts, protocol) {
  const { sid } = opts
  const { id, sdb } = await get(sid)
  const { drive } = sdb

  const by = id
  let db = null
  let msg_id = 0
  let send_to_sidebar = null
  let send_to_newsfeed = null

  const { io, _ } = net_helper(by)
  io.on.write = io_write()

  let active_path = ''
  let active_tab = 'news'

  let newsfeed_el = null
  let write_el = null

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="container">
      <div class="sidebar"></div>
      <div class="main-viewer main"></div>
    </div>
  `

  const sidebar_el = shadow.querySelector('.sidebar')
  const main_viewer = shadow.querySelector('.main-viewer')

  const shell_sheet = new CSSStyleSheet()
  const layout_sheet = new CSSStyleSheet()
  const card_sheet = new CSSStyleSheet()
  const sheet = new CSSStyleSheet()

  shadow.adoptedStyleSheets = [shell_sheet, layout_sheet, card_sheet, sheet]

  const css_files = await Promise.all([
    drive.get('theme/shell.css'),
    drive.get('theme/layout.css'),
    drive.get('theme/news-card.css')
  ])
  const [shell_file, layout_file, card_file] = css_files

  if (shell_file && shell_file.raw) shell_sheet.replaceSync(shell_file.raw)
  if (layout_file && layout_file.raw) layout_sheet.replaceSync(layout_file.raw)
  if (card_file && card_file.raw) card_sheet.replaceSync(card_file.raw)

  const subs = await sdb.watch(onbatch)

  if (subs && subs.length > 0) {
    const menu_el = await menu_sidebar({ sid: subs[0].sid }, sidebar_protocol)
    sidebar_el.appendChild(menu_el)
  }

  newsfeed_el = await newsfeed_view({ sid: subs[1].sid }, newsfeed_protocol)
  write_el = await write_page({ sid: subs[2].sid }, io.invite('write', { news: by }))

  render_main_view()

  shadow.addEventListener('click', handle_shadow_click)
  shadow.addEventListener('submit', handle_shadow_submit)
  shadow.addEventListener('input', handle_shadow_input)

  return el

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(fetch_path_raw_data))
      const valid_data = data.filter(filter_valid_data)
      if (valid_data.length > 0) {
        if (type === 'theme') inject(valid_data)
        if (type === 'entries') on_entries(valid_data)
      }
    }

    function fetch_path_raw_data (path) {
      return drive.get(path).then(extract_raw_if_exists)
    }

    function extract_raw_if_exists (file) {
      return file ? file.raw : null
    }

    function filter_valid_data (d) {
      return d !== null
    }
  }

  function inject (data) {
    if (Array.isArray(data)) {
      sheet.replaceSync(data.join('\n'))
    }
  }

  function sidebar_protocol (send) {
    send_to_sidebar = send
    return on_sidebar_message
  }

  function on_sidebar_message (msg) {
    const { type, data } = msg
    if (type === 'selection_changed') {
      const { selected } = data
      if (selected && selected.length > 0) {
        const path = normalize_path(selected[0])
        if (path !== active_path) {
          active_path = path
          const node_type = db ? db.type(path) : null
          if (node_type === 'subscribed-blog') active_tab = 'blog'
          else active_tab = 'news'
          render_content(path)
        }
      }
    }
    if (type.startsWith('db_')) {
      handle_db_request(msg, send_to_sidebar)
    }
  }

  function newsfeed_protocol (child_handler) {
    send_to_newsfeed = child_handler
    return on_newsfeed_message
  }

  function on_newsfeed_message (msg) {
    const { type, data } = msg
    if (type === 'select_node') {
      render_article_view(data.instance_path)
    }
    if (type === 'navigate_back') {
      render_content(active_path)
    }
  }

  function io_write () {
    const on = {}
    return protocol
    function protocol (m) { (on[m.type] || on_fail)(m) }
    function on_fail () { }
  }

  function handle_shadow_click (e) {
    if (e.target.closest('.news-fab')) {
      render_write_page_view(active_path)
    }
    const tab_target = e.target.closest('.tab')
    if (tab_target) {
      handle_tab_click(tab_target.textContent.trim().toLowerCase())
    }
  }

  function handle_tab_click (tab) {
    if (tab && tab !== active_tab) {
      active_tab = tab
      render_content(active_path)
    }
  }

  function handle_shadow_submit (e) {
    if (e.target.id === 'write-story-form') {
      e.preventDefault()
      const create_story_data = {
        title: e.target.title.value,
        content: e.target.content.value,
        blog: e.target.blog.value
      }
      handle_publish(create_story_data)
    }
  }

  function handle_shadow_input (e) {
    if (e.target.classList.contains('input-content')) {
      handle_content_input(e.target.value)
    }
  }

  function handle_content_input (val) {
    const words = val.trim() === '' ? 0 : val.trim().split(/\s+/).length
    const wc = shadow.getElementById('word-count-span')
    const rt = shadow.getElementById('read-time-span')
    if (wc) wc.textContent = `${words} words`
    if (rt) rt.textContent = `~${Math.ceil(words / 200)} min read`
  }

  function render_main_view () {
    main_viewer.innerHTML = `
      <div class="empty-container">
        <h2 class="empty-title">Select a blog to start</h2>
        <p class="empty-hint">← Explore your blogs and subscriptions in the sidebar</p>
      </div>
    `
  }

  function handle_publish (data) {
    const new_story = {
      title: data.title,
      content: data.content,
      author: 'You',
      date: new Date().toLocaleDateString(),
      description: data.content.slice(0, 100) + '...',
      tags: ['#local', '#' + data.blog.replace(/\s+/g, '')],
      color: '#6366f1'
    }
    save_local_story(new_story)
    alert('Story published locally!')
    render_content(active_path)
  }

  async function on_entries (data) {
    if (!data || !data[0]) {
      db = graphdb({})
      notify_db_initialized({})
      return
    }
    let parsed_data = {}
    try {
      if (typeof data[0] === 'string') {
        parsed_data = JSON.parse(data[0])
      } else if (data[0].raw && typeof data[0].raw === 'string') {
        parsed_data = JSON.parse(data[0].raw)
      } else {
        parsed_data = data[0]
      }
    } catch (e) { }
    db = graphdb(parsed_data)
    notify_db_initialized(parsed_data)
  }

  function notify_db_initialized (entries) {
    if (send_to_sidebar) {
      const head = [by, 'menu_sidebar', msg_id++]
      send_to_sidebar({ head, type: 'db_initialized', data: { entries } })
    }
  }

  async function render_content (path) {
    const node_type = db ? db.type(path) : null
    const node = db ? db.get(path) : null

    if (!node) return render_main_view()

    let tabs_html = ''
    if (node_type !== 'discover') {
      tabs_html = `
        <div class="view-tabs">
          <button class="tab ${active_tab === 'news' ? 'active' : ''}">News</button>
          <button class="tab ${active_tab === 'blog' ? 'active' : ''}">Blog</button>
        </div>
      `
    }

    main_viewer.innerHTML = `
      <div class="content-container">
        ${tabs_html}
        <div class="tab-content"></div>
      </div>
    `
    const content_el = main_viewer.querySelector('.tab-content')

    if (active_tab === 'news') {
      await render_news_tab(path, content_el)
    } else {
      await render_blog_tab(path, content_el)
    }
  }

  async function render_news_tab (path, target) {
    const node = db.get(path)
    const node_subs = node.subs || []
    const fetched_items = []

    for (const sub_path of node_subs) {
      const sub_node = db.get(sub_path)
      if (sub_node && sub_node.posts) {
        for (const post_path of sub_node.posts) {
          const post_data = await fetch_post_data(post_path)
          if (post_data) fetched_items.push({ path: post_path, data: post_data })
        }
      }
    }

    if (node.posts) {
      for (const post_path of node.posts) {
        const post_data = await fetch_post_data(post_path)
        if (post_data) fetched_items.push({ path: post_path, data: post_data })
      }
    }

    if (fetched_items.length > 0) {
      render_feed_view({ items: fetched_items, folder_name: node.name, target: target })
    } else {
      target.innerHTML = '<div class="empty-container"><p>No news to show from subscriptions.</p></div>'
    }
  }

  async function render_blog_tab (path, target) {
    const node = db.get(path)
    const node_type = db.type(path)
    let fetched_items = []

    if (node_type === 'my-blog') {
      const local_stories = get_local_stories()
      fetched_items = local_stories.map(map_local_story)
    } else if (node.posts) {
      for (const post_path of node.posts) {
        const post_data = await fetch_post_data(post_path)
        if (post_data) fetched_items.push({ path: post_path, data: post_data })
      }
    }

    if (fetched_items.length > 0 || node_type === 'my-blog') {
      render_feed_view({ items: fetched_items, folder_name: node.name, target: target })
      if (node_type === 'my-blog') {
        target.insertAdjacentHTML('beforeend', '<div class="news-fab">+</div>')
      }
    } else {
      target.innerHTML = '<div class="empty-container"><p>This blog has not posted anything yet.</p></div>'
    }
  }

  function render_feed_view ({ items, folder_name, target }) {
    target.appendChild(newsfeed_el)
    if (send_to_newsfeed) {
      send_to_newsfeed({
        type: 'render_feed',
        data: { items: items, folder_name: folder_name, is_news_feed: true }
      })
    }
  }

  async function render_article_view (post_path) {
    const post_data = await fetch_post_data(post_path)
    if (!post_data) {
      main_viewer.innerHTML = '<div class="empty-container"><p>Content not available.</p></div>'
      return
    }
    main_viewer.innerHTML = ''
    main_viewer.appendChild(newsfeed_el)
    if (send_to_newsfeed) {
      send_to_newsfeed({
        type: 'render_article',
        data: {
          data: {
            title: post_data.title,
            author: post_data.author,
            date: post_data.date,
            content: post_data.content
          }
        }
      })
    }
  }

  function render_write_page_view (path) {
    const node = db ? db.get(path) : null
    const blog_name = node ? node.name : 'Main Blog'
    _.write('provision', {}, { selected_blog: blog_name })
    main_viewer.innerHTML = ''
    main_viewer.appendChild(write_el)
  }

  function map_local_story (s) {
    return { path: `local-${Math.random()}`, data: s }
  }

  async function fetch_post_data (post_path) {
    try {
      const file = await drive.get(post_path)
      if (file && file.raw) {
        const parsed = await content_parser({ raw: file.raw })
        return parsed
      }
    } catch (e) { }
    return null
  }

  function normalize_path (path) {
    if (!path) return '/'
    if (typeof path !== 'string') return '/'
    return path.split('|').pop() || '/'
  }

  function get_local_stories () {
    try {
      const stories = localStorage.getItem('p2p_stories')
      return stories ? JSON.parse(stories) : []
    } catch (e) { return [] }
  }

  function save_local_story (story) {
    const stories = get_local_stories()
    stories.unshift(story)
    localStorage.setItem('p2p_stories', JSON.stringify(stories))
  }

  async function handle_db_request (request_msg, send) {
    const { head: request_head, type: operation, data: params } = request_msg
    let result
    if (!db) { send_response(request_head, null); return }

    if (operation === 'db_get') result = db.get(params.path)
    else if (operation === 'db_has') result = db.has(params.path)
    else if (operation === 'db_is_empty') result = db.is_empty()
    else if (operation === 'db_root') result = db.root()
    else if (operation === 'db_keys') result = db.keys()
    else if (operation === 'db_raw') result = db.raw()
    else result = null

    send_response(request_head, result)

    function send_response (request_head, result) {
      const response_head = [by, 'menu_sidebar', msg_id++]
      send({ head: response_head, refs: { cause: request_head }, type: 'db_response', data: { result } })
    }
  }
}

function fallback_module () {
  const _ = {
    menu_sidebar: { $: '' },
    newsfeed_view: { $: '' },
    write_page: { $: '' },
    './graphdb': { $: '' },
    'newsfeed_view/content_parser': { $: '' },
    net_helper: { $: '' }
  }

  return { _, api: fallback_instance }

  function fallback_instance (opts) {
    const _ = {
      menu_sidebar: {
        0: '',
        mapping: {
          style: 'style',
          entries: 'entries',
          runtime: 'runtime',
          mode: 'mode',
          flags: 'flags',
          keybinds: 'keybinds',
          undo: 'undo',
          data: 'data'
        }
      },
      newsfeed_view: {
        0: '',
        mapping: { runtime: 'runtime', theme: 'theme' }
      },
      write_page: {
        0: '',
        mapping: { runtime: 'runtime', theme: 'theme' }
      },
      './graphdb': { 0: '' },
      'newsfeed_view/content_parser': { 0: '' }
    }

    const drive = {
      'entries/': { 'entries.json': { $ref: 'entries.json' } },
      'data/': {},
      'news_cards/': {
        'news-card.css': { $ref: '../news_cards/news-card.css' }
      },
      'theme/': {
        'shell.css': {
          raw: `
    body { margin: 0; padding: 0; overflow: hidden; }
    .container {
      display: flex;
      width: 100vw;
      height: 100vh;
      overflow: hidden;
      font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif;
    }
    .sidebar {
      width: 250px;
      min-width: 250px;
      border-right: 1px solid #e5e7eb;
      background: #f9fafb;
      display: flex;
      flex-direction: column;
    }
    .main {
      flex: 1;
      overflow: auto;
      padding: 0;
      background: #ffffff;
    }
          `
        },
        'layout.css': {
          $ref: 'layout.css'
        },
        'news-card.css': {
          $ref: '../news_cards/news-card.css'
        }
      },
      'style/': {},
      'runtime/': {
        'viewer_data.json': { raw: '{}' },
        'write_data.json': { raw: '{}' }
      },
      'mode/': {},
      'flags/': {},
      'keybinds/': {},
      'undo/': {},
      'posts/': {
        'tech-thoughts-1': { $ref: 'data/tech-thoughts-1.md' },
        'tech-thoughts-2': { $ref: 'data/tech-thoughts-2.md' },
        'p2p-experiments-1': { $ref: 'data/p2p-experiments-1.md' },
        'p2p-experiments-2': { $ref: 'data/p2p-experiments-2.md' },
        'security-notes-1': { $ref: 'data/security-notes-1.md' },
        'hackers-digest-1': { $ref: 'data/hackers-digest-1.md' },
        'hackers-digest-2': { $ref: 'data/hackers-digest-2.md' },
        'off-the-grid-1': { $ref: 'data/off-the-grid-1.md' },
        'off-the-grid-2': { $ref: 'data/off-the-grid-2.md' },
        'peer-review-1': { $ref: 'data/peer-review-1.md' },
        'cryptography-weekly-1': { $ref: 'data/cryptography-weekly-1.md' },
        'random-blog-a-1': { $ref: 'data/random-blog-a-1.md' },
        'random-blog-b-1': { $ref: 'data/random-blog-b-1.md' },
        'random-blog-c-1': { $ref: 'data/random-blog-c-1.md' }
      }
    }

    return { _, drive }
  }
}
