// news/index.js
const STATE = require('STATE')
const statedb = STATE(__filename)

const { get } = statedb(fallback_module)
const menu_sidebar = require('menu_sidebar')
const newsfeed_view = require('newsfeed_view')
const content_parser = require('newsfeed_view/content_parser')
const write_page = require('write_page')
const graphdb = require('./graphdb')

module.exports = news_app

async function news_app (opts, protocol) {
  const { sid } = opts
  const { id, sdb } = await get(sid)
  const { drive } = sdb

  const by = id
  let db = null
  let mid = 0
  let send_to_sidebar = null

  let active_folder = ''

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'open' })

  shadow.innerHTML = `
    <div class="container">
      <div class="sidebar"></div>
      <div class="main-viewer main"></div>
    </div>
  `

  const sidebar_el = shadow.querySelector('.sidebar')
  const main_viewer = shadow.querySelector('.main-viewer')

  const shell_sheet = new CSSStyleSheet()
  const layout_sheet = new CSSStyleSheet()
  const sheet = new CSSStyleSheet()

  shadow.adoptedStyleSheets = [shell_sheet, layout_sheet, sheet]

  if (drive) {
    drive.get('theme/shell.css').then(apply_shell_css).catch(ignore_css_error)
    drive.get('theme/layout.css').then(apply_layout_css).catch(ignore_css_error)
  }

  const subs = await sdb.watch(onbatch)

  if (subs && subs.length > 0) {
    const sidebar_instance = subs[0]
    const menu_el = await menu_sidebar({ sid: sidebar_instance.sid }, sidebar_protocol)
    sidebar_el.appendChild(menu_el)
  }

  render_main_view()

  shadow.addEventListener('click', handle_shadow_click)
  shadow.addEventListener('submit', handle_shadow_submit)
  shadow.addEventListener('input', handle_shadow_input)

  return el

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(fetch_path_raw_data))
      const valid_data = data.filter(filter_valid_data)
      if (valid_data.length > 0) {
        if (type === 'theme') inject(valid_data)
        if (type === 'entries') on_entries(valid_data)
      }
    }

    function fetch_path_raw_data (path) {
      return drive.get(path).then(extract_raw_if_exists)
    }

    function extract_raw_if_exists (file) {
      return file ? file.raw : null
    }

    function filter_valid_data (d) {
      return d !== null
    }
  }

  function inject (data) {
    if (Array.isArray(data)) {
      sheet.replaceSync(data.join('\n'))
    }
  }

  function sidebar_protocol (send) {
    send_to_sidebar = send
    return on_sidebar_message
  }

  function on_sidebar_message (msg) {
    const { type, data } = msg
    if (type === 'selection_changed') {
      const { selected } = data
      if (selected && selected.length > 0) render_folder_content(selected[0])
    }
    if (type.startsWith('db_')) {
      handle_db_request(msg, send_to_sidebar)
    }
  }

  function apply_shell_css (file) {
    if (file && file.raw) shell_sheet.replaceSync(file.raw)
  }

  function apply_layout_css (file) {
    if (file && file.raw) layout_sheet.replaceSync(file.raw)
  }

  function ignore_css_error (e) { }

  function handle_shadow_click (e) {
    if (e.target.closest('.news-fab')) {
      render_write_page_view(active_folder)
    }
  }

  function handle_shadow_submit (e) {
    if (e.target.id === 'write-story-form') {
      e.preventDefault()
      const create_story_data = {
        title: e.target.title.value,
        content: e.target.content.value,
        blog: e.target.blog.value
      }
      handle_publish(create_story_data)
    }
  }

  function handle_shadow_input (e) {
    if (e.target.classList.contains('input-content')) {
      const val = e.target.value
      const words = val.trim() === '' ? 0 : val.trim().split(/\s+/).length
      const wc = shadow.getElementById('word-count-span')
      const rt = shadow.getElementById('read-time-span')
      if (wc) wc.textContent = `${words} words`
      if (rt) rt.textContent = `~${Math.ceil(words / 200)} min read`
    }
  }

  function render_main_view () {
    const html = `
      <div class="empty-container">
        <h2 class="empty-title">Select an item to read</h2>
        <p class="empty-hint">← Choose a story or feed from the sidebar</p>
      </div>
    `
    main_viewer.innerHTML = html
  }

  async function render_article (data) {
    await sdb.drive.put('runtime/viewer_data.json', JSON.stringify({ data }))
    const viewer_sid = [...id, 'newsfeed_view', mid++]
    const element = await newsfeed_view({ sid: viewer_sid }, newsfeed_protocol)
    main_viewer.innerHTML = ''
    main_viewer.appendChild(element)
  }

  function newsfeed_protocol (send) {
    return on_newsfeed_message
  }

  function on_newsfeed_message (msg) {
    const { type, data } = msg
    if (type === 'select_node') {
      render_folder_content(data.instance_path)
    }
  }

  async function render_write_page_view (folder_name) {
    const html = await write_page({})
    main_viewer.innerHTML = ''
    if (typeof html === 'string') {
      main_viewer.innerHTML = html
    } else {
      main_viewer.appendChild(html)
    }
  }

  function handle_publish (data) {
    const new_story = {
      title: data.title,
      content: data.content,
      author: 'You',
      date: new Date().toLocaleDateString(),
      description: data.content.slice(0, 100) + '...',
      tags: ['#local', '#' + data.blog.replace(/\s+/g, '')],
      color: '#6366f1'
    }
    save_local_story(new_story)
    alert('Story published locally!')
    render_folder_content('my-stories')
  }

  async function on_entries (data) {
    if (!data || !data[0]) {
      db = graphdb({})
      notify_db_initialized({})
      return
    }
    let parsed_data = {}
    try {
      if (typeof data[0] === 'string') {
        parsed_data = await content_parser({ raw: data[0] }) || {}
      } else {
        parsed_data = data[0]
      }
    } catch (e) { }
    db = graphdb(parsed_data)
    notify_db_initialized(parsed_data)
  }

  function notify_db_initialized (entries) {
    if (send_to_sidebar) {
      const head = [by, 'menu_sidebar', mid++]
      send_to_sidebar({ head, type: 'db_initialized', data: { entries } })
    }
  }

  async function render_folder_content (path) {
    let drive_path = path
    if (drive_path.startsWith('/')) drive_path = drive_path.slice(1)

    const data = await try_get_folder_data(drive_path)

    if (data && data.content) {
      render_article(data)
    } else {
      const { folder_name, subs } = get_folder_details(path)

      if (subs.length > 0) {
        const handled = await process_folder_items({ path, subs, folder_name })
        if (handled) return
      }

      render_empty_folder(folder_name, path)
    }
  }

  async function try_get_folder_data (drive_path) {
    try {
      const file = await drive.get(drive_path)
      return await content_parser({ raw: file.raw })
    } catch (e) {
      return null
    }
  }

  function get_folder_details (path) {
    let folder_name = path.split('/').pop() || path
    let subs = []
    let db_path = path
    if (typeof path === 'string' && path.includes('|')) {
      db_path = path.split('|').pop()
    }

    if (db) {
      const entry = db.get(db_path) || db.get(path)
      if (entry) {
        if (entry.name) folder_name = entry.name
        if (entry.subs) subs = entry.subs
      }
    }
    return { folder_name, subs }
  }

  async function process_folder_items (opts) {
    const { path, subs, folder_name } = opts
    console.log('[news] process_folder_items:', { path, subs, folder_name })
    active_folder = folder_name
    const fetched_items = (await Promise.all(subs.map(fetch_sub_item))).filter(filter_valid_fetched_items)
    console.log('[news] fetched_items count:', fetched_items.length)

    async function fetch_sub_item (sub_path) {
      console.log('[news] fetch_sub_item:', sub_path)
      let item_data
      try {
        let sub_drive_path = sub_path
        if (sub_drive_path.startsWith('/')) sub_drive_path = sub_drive_path.slice(1)

        const file = await drive.get(sub_drive_path)
        console.log('[news] drive.get result for:', sub_drive_path, file ? 'Found' : 'Missing')
        if (file) {
          item_data = await content_parser({ raw: file.raw })
          console.log('[news] content_parser result for:', sub_drive_path, item_data ? 'Success' : 'Failed')
        }
      } catch (e) {
        console.error('[news] Error fetching sub item:', sub_path, e)
      }
      if (!item_data) return null
      return { path: sub_path, data: item_data }
    }

    function filter_valid_fetched_items (item) {
      return item && item.data
    }

    if (path.includes('my-stories')) {
      const local_stories = get_local_stories()
      local_stories.forEach(push_local_story)

      function push_local_story (story) {
        fetched_items.push({ path: 'local-' + Date.now() + Math.random(), data: story })
      }
    }

    if (fetched_items.length > 0) {
      const is_my_stories = path.includes('my-stories') || path.includes('My Stories')

      const view_data = {
        items: fetched_items,
        folder_name,
        is_my_stories
      }
      await sdb.drive.put('runtime/viewer_data.json', JSON.stringify({ data: view_data }))

      const viewer_sid = [...id, 'newsfeed_view', mid++]
      const element = await newsfeed_view({ sid: viewer_sid }, newsfeed_protocol)

      main_viewer.innerHTML = ''
      main_viewer.appendChild(element)
      return true
    }
    return false
  }

  function render_empty_folder (folder_name, path) {
    active_folder = folder_name
    const html = `
      <div class="empty-folder-container">
        <div class="empty-folder-icon">📂</div>
        <h2 class="empty-folder-title">${folder_name}</h2>
        <p>Select a file inside to view content.</p>
        ${(path.includes('my-stories') || folder_name.includes('My Stories')) ? '<div class="news-fab">+</div>' : ''}
      </div>
    `
    main_viewer.innerHTML = html
  }

  function get_local_stories () {
    try {
      const stories = localStorage.getItem('p2p_stories')
      return stories ? JSON.parse(stories) : []
    } catch (e) { return [] }
  }

  function save_local_story (story) {
    const stories = get_local_stories()
    stories.unshift(story)
    localStorage.setItem('p2p_stories', JSON.stringify(stories))
  }

  async function handle_db_request (request_msg, send) {
    const { head: request_head, type: operation, data: params } = request_msg
    let result
    if (!db) { send_response(request_head, null); return }

    if (operation === 'db_get') result = db.get(params.path)
    else if (operation === 'db_has') result = db.has(params.path)
    else if (operation === 'db_is_empty') result = db.is_empty()
    else if (operation === 'db_root') result = db.root()
    else if (operation === 'db_keys') result = db.keys()
    else if (operation === 'db_raw') result = db.raw()
    else result = null

    send_response(request_head, result)

    function send_response (request_head, result) {
      const response_head = [by, 'menu_sidebar', mid++]
      send({ head: response_head, refs: { cause: request_head }, type: 'db_response', data: { result } })
    }
  }
}

function fallback_module () {
  const _ = {
    menu_sidebar: { $: '' },
    './graphdb': { $: '' },
    newsfeed_view: { $: '' },
    'newsfeed_view/content_parser': { $: '' },
    write_page: { $: '' }
  }

  return { _, api: fallback_instance }

  function fallback_instance (opts) {
    const _ = {
      menu_sidebar: {
        0: '',
        mapping: {
          style: 'style',
          entries: 'entries',
          runtime: 'runtime',
          mode: 'mode',
          flags: 'flags',
          keybinds: 'keybinds',
          undo: 'undo'
        }
      },
      './graphdb': { 0: '' },
      newsfeed_view: {
        0: '',
        '*': '',
        mapping: { runtime: 'runtime', theme: 'theme' }
      },
      'newsfeed_view/content_parser': { 0: '' },
      write_page: { 0: '' }
    }

    const drive = {
      'entries/': { 'entries.json': { $ref: 'entries.json' } },
      'theme/': {
        'shell.css': {
          raw: `
    body { margin: 0; padding: 0; overflow: hidden; }
    .container {
      display: flex;
      width: 100vw;
      height: 100vh;
      overflow: hidden;
      font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif;
    }
    .sidebar {
      width: 250px;
      min-width: 250px;
      border-right: 1px solid #e5e7eb;
      background: #f9fafb;
      display: flex;
      flex-direction: column;
    }
    .main {
      flex: 1;
      overflow: auto;
      padding: 2rem;
      background: #ffffff;
    }
          `
        },
        'layout.css': {
          $ref: 'layout.css'
        },
        'news-card.css': {
          $ref: '../news_cards/news-card.css'
        }
      },
      'style/': {},
      'runtime/': {
        'viewer_data.json': { raw: '{}' }
      },
      'mode/': {},
      'flags/': {},
      'keybinds/': {},
      'undo/': {},
      'feeds/hackers-digest/code-coffee': { $ref: 'data/code-coffee.md' },
      'feeds/hackers-digest/system-design': { $ref: 'data/system-design.md' },
      'feeds/off-the-grid/mesh-network': { $ref: 'data/mesh-network.md' },
      'feeds/off-the-grid/fediverse': { $ref: 'data/fediverse.md' },
      'feeds/off-the-grid/self-hosting': { $ref: 'data/self-hosting.md' },
      'feeds/peer-review/network-notes': { $ref: 'data/network-notes.md' },
      'feeds/peer-review/security-chronicles/privacy-matters': { $ref: 'data/privacy-matters.md' },
      'feeds/peer-review/security-chronicles/zero-trust': { $ref: 'data/zero-trust.md' },
      'lists/best-of-tech': { $ref: 'data/best-of-tech.md' },
      'lists/morning-read': { $ref: 'data/morning-read.md' },
      'my-stories/story-1': { $ref: 'data/story-1.md' },
      'my-stories/story-2': { $ref: 'data/story-2.md' },
      'my-stories/story-3': { $ref: 'data/story-3.md' },
      'my-stories/story-4': { $ref: 'data/story-4.md' }
    }

    return { _, drive }
  }
}
