// news/wrapper.js
const STATE = require('STATE')
console.log('[DEBUG] wrapper.js running, filename:', __filename)
const statedb = STATE(__filename)

const { get } = statedb(fallback_module)
const graph_explorer = require('graph-explorer')
const graphdb = require('./graphdb')

const dummy_content = require('./dummy_content')

module.exports = my_component_with_graph

async function my_component_with_graph(opts, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const by = id
  let db = null
  let send_to_graph_explorer = null
  let mid = 0

  const on = {
    theme: inject,
    entries: on_entries
  }

  const el = document.createElement('div')
  el.style.position = 'absolute'
  el.style.top = '0'
  el.style.left = '0'
  el.style.right = '0'
  el.style.bottom = '0'
  el.style.display = 'flex' // Flexbox layout

  const shadow = el.attachShadow({ mode: 'closed' })
  const sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [sheet]

  const subs = await sdb.watch(onbatch)
  const explorer_el = await graph_explorer(subs[0], graph_explorer_protocol)


  explorer_el.style.width = '250px'
  explorer_el.style.flexShrink = '0'
  explorer_el.style.borderRight = '1px solid #e5e7eb'


  const main_content = document.createElement('div')
  main_content.style.flex = '1'
  main_content.style.padding = '40px'
  main_content.style.overflowY = 'auto'
  main_content.style.backgroundColor = '#ffffff'
  main_content.style.color = '#1f2937'
  main_content.style.fontFamily = "'Inter', sans-serif"
  main_content.innerHTML = `
    <div style="text-align: center; margin-top: 100px; color: #9ca3af;">
      <h2>Select an item to read</h2>
      <p>← Choose a story or feed from the sidebar</p>
    </div>
  `

  shadow.append(explorer_el)
  shadow.append(main_content)



  explorer_el.addEventListener('click', (e) => {
    const nodeEl = e.target.closest('.node')
    if (!nodeEl) return


    if (e.target.closest('.name') || e.target.closest('.prefix')) return


    const instance_path = nodeEl.dataset.instance_path
    if (instance_path && send_to_graph_explorer) {
      console.log('[Interaction] Whitespace/Icon toggle:', instance_path)
      const head = [id, 'graph_explorer', mid++]
      send_to_graph_explorer({
        head,
        type: 'toggle_node',
        data: { instance_path, toggle_type: 'subs' }
      })
    }
  })


  window.addEventListener('news-select', (e) => {
    const instance_path = e.detail
    if (instance_path && send_to_graph_explorer) {
      console.log('[Interaction] List Item Click:', instance_path)
      const head = [id, 'graph_explorer', mid++]


      send_to_graph_explorer({
        head,
        type: 'select_node',
        data: { instance_path }
      })

    }
  })

  return el

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file ? file.raw : null)))
      const valid_data = data.filter(d => d !== null)
      if (valid_data.length > 0) {
        on[type] && on[type](valid_data)
      }
    }
  }

  function inject(data) {
    sheet.replaceSync(data.join('\n'))
  }

  function on_entries(data) {
    if (!data || !data[0]) {

      db = graphdb({})
      notify_db_initialized({})
      return
    }

    let parsed_data
    try {
      parsed_data = typeof data[0] === 'string' ? JSON.parse(data[0]) : data[0]
    } catch (e) {
      console.error('Failed to parse entries data:', e)
      parsed_data = {}
    }

    if (typeof parsed_data !== 'object' || !parsed_data) {
      console.error('Parsed entries data is not a valid object.')
      parsed_data = {}
    }

    db = graphdb(parsed_data)
    notify_db_initialized(parsed_data)
  }

  function notify_db_initialized(entries) {
    if (send_to_graph_explorer) {
      const head = [by, 'graph_explorer', mid++]
      send_to_graph_explorer({
        head,
        type: 'db_initialized',
        data: { entries }
      })
    }
  }

  function graph_explorer_protocol(send) {
    send_to_graph_explorer = send
    return on_graph_explorer_message

    function on_graph_explorer_message(msg) {
      const { type, data } = msg


      if (type === 'selection_changed') {
        const { selected } = data
        if (selected && selected.length > 0) {
          const path = selected[0]
          console.log('[Sidebar] Selection Changed:', path)
          render_content(path)
        }
      }


      if (type === 'confirmed_nodes') {
      }

      if (type.startsWith('db_')) {
        handle_db_request(msg, send)
      }
    }

    function render_content(path) {
      let data = dummy_content[path]

      if (!data) {
        const key = Object.keys(dummy_content).find(k => path.includes(k) || path.endsWith(k))
        if (key) data = dummy_content[key]
      }


      if (data && data.content && data.content.trim().length > 0) {
        main_content.innerHTML = `
          <article style="max-width: 800px; margin: 0 auto;">
            <header style="margin-bottom: 2rem; border-bottom: 1px solid #eee; padding-bottom: 1rem;">
              <h1 style="font-size: 2.5rem; font-weight: 700; color: #111827; margin-bottom: 0.5rem;">${data.title}</h1>
              <div style="color: #6b7280; font-size: 0.9rem;">
                <span>By <strong>${data.author}</strong></span> • <span>${data.date}</span>
              </div>
            </header>
            <div style="font-size: 1.1rem; line-height: 1.7; color: #374151;">
              ${data.content}
            </div>
          </article>
        `
      } else {

        let folderName = path
        let subs = []


        // Instance path format: |/|/grandparent|/parent|/node
        let dbPath = path
        if (typeof path === 'string' && path.includes('|')) {
          dbPath = path.split('|').pop()
        }

        if (db) {

          const entry = db.get(dbPath)
          if (entry) {
            if (entry.name) folderName = entry.name
            if (entry.subs) subs = entry.subs
          } else {

            const rawEntry = db.get(path)
            if (rawEntry) {
              if (rawEntry.name) folderName = rawEntry.name
              if (rawEntry.subs) subs = rawEntry.subs
            }
          }
        }


        if (subs.length > 0) {
          const listItems = subs.map(subPath => {

            let itemData = dummy_content[subPath]
            if (!itemData) {
              const key = Object.keys(dummy_content).find(k => subPath.includes(k) || subPath.endsWith(k))
              if (key) itemData = dummy_content[key]
            }
            return { path: subPath, data: itemData }
          }).filter(item => item.data)

          if (listItems.length > 0) {
            const isMyStories = path.includes('/my-stories')


            const filtersHtml = ''


            const fabHtml = isMyStories ? `
               <button style="position: fixed; bottom: 2rem; right: 2rem; width: 56px; height: 56px; border-radius: 50%; background-color: #6366f1; color: white; border: none; box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06); cursor: pointer; display: flex; align-items: center; justify-content: center; font-size: 2rem; transition: transform 0.2s;"
                       onmouseover="this.style.transform='scale(1.1)'" onmouseout="this.style.transform='scale(1)'">
                 +
               </button>
             ` : ''

            const cardsHtml = listItems.map(({ path, data }) => {


              const metaTop = isMyStories ? `
                   <div style="display: flex; align-items: center; gap: 0.5rem; font-size: 0.875rem; color: #6b7280; margin-bottom: 0.25rem;">
                     <span style="font-weight: 600; color: #111827;">${data.author}</span>
                     <span>•</span>
                     <span>${data.date}</span>
                     <span>•</span>
                     <span>${Math.floor(Math.random() * 10) + 2} min</span>
                   </div>
                ` : ''

              const metaBottom = !isMyStories ? `
                   <div style="display: flex; align-items: center; gap: 0.75rem; font-size: 0.875rem; color: #6b7280; margin-top: 0.5rem;">
                     <span style="font-weight: 600; color: #6b7280;">${data.author}</span>
                     <span>•</span>
                     <span>${data.date}</span>
                     ${data.tags ? data.tags.map(tag => `<span style="color: #6366f1; background: #e0e7ff; padding: 2px 10px; border-radius: 4px; font-size: 0.75rem; font-weight: 500;">${tag}</span>`).join('') : ''}
                   </div>
                ` : ''

              const tagsBottom = isMyStories ? `
                   <div style="display: flex; align-items: center; gap: 0.5rem;">
                     ${data.tags ? data.tags.map(tag => `<span style="color: #6366f1; background: #e0e7ff; padding: 2px 10px; border-radius: 4px; font-size: 0.75rem; font-weight: 500;">${tag}</span>`).join('') : ''}
                   </div>
                ` : ''

              return `
               <div onclick="window.dispatchEvent(new CustomEvent('news-select', { detail: '${path}' }))" 
                    style="display: flex; gap: 1.5rem; padding: 1.5rem 0; border-bottom: 1px solid #f3f4f6; cursor: pointer; transition: background 0.2s;"
                    onmouseover="this.style.background='rgba(0,0,0,0.01)'" 
                    onmouseout="this.style.background='transparent'">
                 
                 <!-- Avatar/Icon -->
                 <div style="flex-shrink: 0; width: 48px; height: 48px; border-radius: 50%; background-color: ${data.color || '#e5e7eb'}; display: flex; align-items: center; justify-content: center; font-size: 1.25rem; color: white; font-weight: 600;">
                    ${data.title.charAt(0)}
                 </div>

                 <!-- Content -->
                 <div style="flex: 1;">
                   ${metaTop}
                   
                   <h3 style="font-size: 1.25rem; font-weight: 700; color: #111827; margin: 0 0 0.5rem 0;">${data.title}</h3>
                   <p style="color: #4b5563; font-size: 1rem; line-height: 1.5; margin: 0 0 0.5rem 0;">
                     ${data.description || 'No description available.'}
                   </p>
                   
                   ${metaBottom}
                   ${tagsBottom}
                 </div>
               </div>
             `}).join('')

            main_content.innerHTML = `
               <div style="max-width: 900px; margin: 0 auto; padding-bottom: 80px;">
                 <header style="margin-bottom: 2rem; display: flex; justify-content: space-between; align-items: flex-start;">
                   <div>
                     <h2 style="font-size: 2rem; font-weight: 700; color: #111827; margin: 0;">${folderName}</h2>
                     ${isMyStories ? '<p style="color: #6b7280; margin-top: 0.5rem;">Your published posts</p>' : ''}
                   </div>
                   ${''}
                 </header>
                 
                 ${filtersHtml}
                 
                 <div>
                   ${cardsHtml}
                 </div>
               </div>
               ${fabHtml}
             `

            return
          }
        }


        main_content.innerHTML = `
          <div style="text-align: center; margin-top: 50px; color: #9ca3af;">
            <div style="font-size: 4rem; opacity: 0.2; margin-bottom: 1rem;">📂</div>
            <h2 style="color: #374151;">${folderName}</h2>
            <p>Select a file inside to view content.</p>
          </div>
        `
      }
    }

    function render_card(path, data, isMyStories) {
      const metaTop = isMyStories ? `
         <div style="display: flex; align-items: center; gap: 0.5rem; font-size: 0.875rem; color: #6b7280; margin-bottom: 0.25rem;">
           <span style="font-weight: 600; color: #111827;">${data.author}</span>
           <span>•</span>
           <span>${data.date}</span>
           <span>•</span>
           <span>${Math.floor(Math.random() * 10) + 2} min</span>
         </div>
      ` : ''

      const metaBottom = !isMyStories ? `
         <div style="display: flex; align-items: center; gap: 0.75rem; font-size: 0.875rem; color: #6b7280; margin-top: 0.5rem;">
           <span style="font-weight: 600; color: #6b7280;">${data.author}</span>
           <span>•</span>
           <span>${data.date}</span>
           ${data.tags ? data.tags.map(tag => `<span style="color: #6366f1; background: #e0e7ff; padding: 2px 10px; border-radius: 4px; font-size: 0.75rem; font-weight: 500;">${tag}</span>`).join('') : ''}
         </div>
      ` : ''

      const tagsBottom = isMyStories ? `
         <div style="display: flex; align-items: center; gap: 0.5rem;">
           ${data.tags ? data.tags.map(tag => `<span style="color: #6366f1; background: #e0e7ff; padding: 2px 10px; border-radius: 4px; font-size: 0.75rem; font-weight: 500;">${tag}</span>`).join('') : ''}
         </div>
      ` : ''

      return `
     <div onclick="window.dispatchEvent(new CustomEvent('news-select', { detail: '${path}' }))" 
          style="display: flex; gap: 1.5rem; padding: 1.5rem 0; border-bottom: 1px solid #f3f4f6; cursor: pointer; transition: background 0.2s;"
          onmouseover="this.style.background='rgba(0,0,0,0.01)'" 
          onmouseout="this.style.background='transparent'">
       
       <!-- Avatar/Icon -->
       <div style="flex-shrink: 0; width: 48px; height: 48px; border-radius: 50%; background-color: ${data.color || '#e5e7eb'}; display: flex; align-items: center; justify-content: center; font-size: 1.25rem; color: white; font-weight: 600;">
          ${data.title.charAt(0)}
       </div>

       <!-- Content -->
       <div style="flex: 1;">
         ${metaTop}
         
         <h3 style="font-size: 1.25rem; font-weight: 700; color: #111827; margin: 0 0 0.5rem 0;">${data.title}</h3>
         <p style="color: #4b5563; font-size: 1rem; line-height: 1.5; margin: 0 0 0.5rem 0;">
           ${data.description || 'No description available.'}
         </p>
         
         ${metaBottom}
         ${tagsBottom}
       </div>
     </div>
   `
    }

    async function handle_db_request(request_msg, send) {
      const { head: request_head, type: operation, data: params } = request_msg
      let result

      if (!db) {
        // console.error('[my_component] Database not initialized yet')
        send_response(request_head, null)
        return
      }

      if (operation === 'db_get') {
        result = db.get(params.path)
      } else if (operation === 'db_has') {
        result = db.has(params.path)
      } else if (operation === 'db_is_empty') {
        result = db.is_empty()
      } else if (operation === 'db_root') {
        result = db.root()
      } else if (operation === 'db_keys') {
        result = db.keys()
      } else if (operation === 'db_raw') {
        result = db.raw()
      } else {
        console.warn('[my_component] Unknown db operation:', operation)
        result = null
      }

      send_response(request_head, result)

      function send_response(request_head, result) {
        const response_head = [by, 'graph_explorer', mid++]
        send({
          head: response_head,
          refs: { cause: request_head },
          type: 'db_response',
          data: { result }
        })
      }
    }
  }
}

function fallback_module() {
  return {
    _: {
      'graph-explorer': { $: '' },
      './graphdb': { $: '' },
      './dummy_content': { $: '' }
    },
    api: fallback_instance
  }

  function fallback_instance() {
    return {
      _: {
        'graph-explorer': {
          0: override_theme,
          mapping: {
            style: 'theme',
            runtime: 'runtime',
            mode: 'mode',
            flags: 'flags',
            keybinds: 'keybinds',
            undo: 'undo'
          }
        },
        './graphdb': {
          0: ''
        },
        './dummy_content': {
          0: '' // Static load
        }
      },
      drive: {
        'theme/': {
          'style.css': {
            raw: ''
          }
        },
        'entries/': {
          'entries.json': {
            $ref: 'entries.json'
          }
        },
        'runtime/': {},
        'mode/': {},
        'flags/': {
          'hubs.json': { raw: '"default"' },
          'selection.json': { raw: 'true' },
          'recursive_collapse.json': { raw: 'true' }
        },
        'keybinds/': {
          'navigation.json': {
            raw: JSON.stringify({
              ArrowUp: 'navigate_up',
              ArrowDown: 'navigate_down',
              ArrowLeft: 'navigate_left',
              ArrowRight: 'navigate_right',
              ' ': 'toggle_expansion',
              Enter: 'toggle_expansion'
            })
          }
        },
        'undo/': {}
      }
    }
  }

  function override_theme() {
    return {
      _: {
        mapping: {
          style: 'theme',
          runtime: 'runtime',
          mode: 'mode',
          flags: 'flags',
          keybinds: 'keybinds',
          undo: 'undo'
        }
      },
      drive: {
        'style/': {
          'style.css': {
            raw: `
            :host {
              --primary: #6366f1;
              --bg-sidebar: #f9fafb;
              --border-color: #e5e7eb;
              --text-main: #1f2937;
              --text-muted: #6b7280;

              display: block;
              height: 100%;
              width: 100%;
              font-family: 'Inter', -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif;
              color: var(--text-main);
              background-color: var(--bg-sidebar);
            }

            .graph-container {
              /* aside */
              width: 100%;
              height: 100vh;
              background: var(--bg-sidebar);
              border-right: 1px solid var(--border-color);
              padding: 1rem 0;
              display: flex;
              flex-direction: column;
              overflow: auto;
              box-sizing: border-box;
            }

            .node {
              /* .tree-item */
              display: flex;
              align-items: center;
              gap: 0.5rem;
              padding: 0.4rem 0.75rem; /* slightly more horiz padding */
              margin: 0.1rem 0.5rem;   /* spacing between items */
              border-radius: 6px;
              cursor: pointer;
              font-size: 0.9rem;
              color: var(--text-main);
              transition: background 0.15s;
              height: 32px; /* Fixed height for consistency */ 
              box-sizing: border-box;
              user-select: none;
            }

            .clickable {
              cursor: pointer;
            }

            .node:hover {
              background: rgba(0, 0, 0, 0.04);
            }

          
            .node > .name:hover {
              color: var(--primary);
            }

            input[type="checkbox"] {
              display: none !important;
            }

       
            .node.type-stories,
            .node.type-feeds,
            .node.type-lists,
            .node.type-discover,
            .node.type-folder {
              font-weight: 600;
              color: #111827; /* Darker text */
            }

       
            .node.type-file, .node.type-js-file {
              font-weight: 400;
              color: var(--text-main);
            }

            /* Map .sidebar-item.active styles */
            .node.selected {
              background: rgba(99, 102, 241, 0.1);
              color: var(--primary);
              font-weight: 600;
            }

  
            .wand {
              display: none !important;
            }
         
            .node.type-root { display: none !important; }


            .icon {
              color: var(--text-muted);
              width: 16px;
              height: 16px;
              display: inline-flex;
              align-items: center;
              justify-content: center;
              margin-right: 0; 
              opacity: 1;      
            }

        
            .prefix {
              cursor: pointer;
              min-width: 16px; 
              height: 100%;
              display: inline-flex;
              align-items: center;
              justify-content: center;
            }

            .indent {
              height: 100%;
              display: inline-flex;
            }
            
       
            .pipe, .blank {
              display: inline-block;
              width: 24px; /* Width per indentation level */
              height: 100%;
            }
            
            
            .prefix::after {
              content: '▸';
              font-size: 10px;
              color: var(--text-muted);
              display: inline-block;
              transition: transform 0.15s;
            }
            .prefix.expanded::after {
               transform: rotate(90deg);
            }
           
            .node.type-stories > .icon::before { content: '📝'; font-size: 14px; }
            .node.type-feeds > .icon::before   { content: '📡'; font-size: 14px; }
            .node.type-lists > .icon::before   { content: '📋'; font-size: 14px; }
            .node.type-discover > .icon::before { content: '🧭'; font-size: 14px; }

      
            .node.type-folder > .icon { 
              display: none;
            }
            
    
            .node.type-file > .icon, .node.type-js-file > .icon {
              display: none;
            }
            
           
            .node.selected .icon {
              color: var(--primary);
            }
            .node.selected.type-file > .icon::before,
            .node.selected.type-js-file > .icon::before {
              background-color: var(--primary);
            }
            

            .graph-container::-webkit-scrollbar {
              width: 6px;
            }
            .graph-container::-webkit-scrollbar-track {
              background: transparent;
            }
            .graph-container::-webkit-scrollbar-thumb {
              background: #d1d5db;
              border-radius: 3px;
            }
            .graph-container::-webkit-scrollbar-thumb:hover {
              background: #9ca3af;
            }
          `
          }
        },
        'runtime/': {
          'node_height.json': { raw: '16' },
          'vertical_scroll_value.json': { raw: '0' },
          'horizontal_scroll_value.json': { raw: '0' },
          'selected_instance_paths.json': { raw: '[]' },
          'confirmed_selected.json': { raw: '[]' },
          'instance_states.json': {
            raw: JSON.stringify({
              '|/': { expanded_subs: true },
              '|/my-stories': { expanded_subs: true },
              '|/feeds': { expanded_subs: true },
              '|/feeds/hackers-digest': { expanded_subs: true },
              '|/feeds/peer-review': { expanded_subs: true },
              '|/feeds/peer-review/security-chronicles': { expanded_subs: true },
              '|/lists': { expanded_subs: true },
              '|/discover': { expanded_subs: true }
            })
          },
          'search_entry_states.json': { raw: '{}' },
          'last_clicked_node.json': { raw: 'null' },
          'view_order_tracking.json': { raw: '{}' }
        },
        'mode/': {
          'current_mode.json': { raw: '"menubar"' },
          'previous_mode.json': { raw: '"menubar"' },
          'search_query.json': { raw: '""' },
          'multi_select_enabled.json': { raw: 'false' },
          'select_between_enabled.json': { raw: 'false' }
        },
        'flags/': {
          'hubs.json': { raw: '"default"' },
          'selection.json': { raw: 'true' },
          'recursive_collapse.json': { raw: 'true' }
        },
        'keybinds/': {
          'navigation.json': {
            raw: JSON.stringify({
              ArrowUp: 'navigate_up_current_node',
              ArrowDown: 'navigate_down_current_node',
              'Control+ArrowDown': 'toggle_subs_for_current_node',
              'Control+ArrowUp': 'toggle_hubs_for_current_node',
              'Alt+s': 'multiselect_current_node',
              'Alt+b': 'select_between_current_node',
              'Control+m': 'toggle_search_mode',
              'Alt+j': 'jump_to_next_duplicate'
            })
          }
        },
        'undo/': {
          'stack.json': { raw: '[]' }
        }
      }
    }
  }
}
