// news/wrapper.js
const STATE = require('STATE')
const statedb = STATE(__filename)

const { get } = statedb(fallback_module)
const graph_explorer = require('graph-explorer')
const article_viewer = require('article-viewer')
const content_parser = article_viewer.parser
const news_card = require('news-card')
const write_page = require('write-page')
const news_list = require('news-list')
const graphdb = require('./graphdb')

module.exports = wrapper

async function wrapper(opts, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const by = id
  let db = null
  let mid = 0
  let send_to_graph_explorer = null
  let explorer_el = null

  const card_map = new WeakMap()
  let list_items_data = []

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  const layout_sheet = new CSSStyleSheet()
  drive.get('theme/layout.css').then(file => {
    layout_sheet.replaceSync(file.raw)
  }).catch(e => console.error('Failed to load layout.css', e))

  const sheet = new CSSStyleSheet()
  const news_card_sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [layout_sheet, sheet]

  const subs = await sdb.watch(onbatch)

  if (subs && subs.length > 0) {
    if (typeof graph_explorer === 'function') {
      explorer_el = await graph_explorer(subs[0], graph_explorer_protocol)
      explorer_el.classList.add('explorer-panel')
      shadow.appendChild(explorer_el)
    }
  }

  render_main_view()

  shadow.addEventListener('click', (e) => {
    const node_el = e.target.closest('.node')
    if (node_el && !e.target.closest('.name') && !e.target.closest('.prefix')) {
      const instance_path = node_el.dataset.instance_path
      if (instance_path && send_to_graph_explorer) {
        const head = [id, 'graph_explorer', mid++]
        send_to_graph_explorer({
          head,
          type: 'toggle_node',
          data: { instance_path, toggle_type: 'subs' }
        })
      }
    }

    const card_target = e.target.closest('.news-card')
    if (card_target) {
      const path = card_map.get(card_target)
      if (path && list_items_data) {
        if (typeof path === 'object' || path.startsWith('local-')) {
          const item = list_items_data.find(i => i.path === path)
          if (item) {
            render_article(item.data)
            return
          }
        }
        const head = [id, 'graph_explorer', mid++]
        send_to_graph_explorer({ head, type: 'select_node', data: { instance_path: path } })
      }
    }

    if (e.target.closest('.news-fab')) {
      const folder_name = e.target.closest('.news-fab').dataset.folder
      render_write_page_view(folder_name)
    }
  })

  shadow.addEventListener('submit', (e) => {
    if (e.target.id === 'write-story-form') {
      e.preventDefault()
      const create_story_data = {
        title: e.target.title.value,
        content: e.target.content.value,
        blog: e.target.blog.value
      }
      handle_publish(create_story_data)
    }
  })

  shadow.addEventListener('input', (e) => {
    if (e.target.classList.contains('input-content')) {
      const val = e.target.value
      const words = val.trim() === '' ? 0 : val.trim().split(/\s+/).length
      const wc = shadow.getElementById('word-count-span')
      const rt = shadow.getElementById('read-time-span')
      if (wc) wc.textContent = `${words} words`
      if (rt) rt.textContent = `~${Math.ceil(words / 200)} min read`
    }
  })

  return el

  function render_html(html_string) {
    let main_viewer = shadow.querySelector('.main-viewer')
    if (!main_viewer) {
      const div = document.createElement('div')
      div.className = 'main-viewer'
      shadow.appendChild(div)
      main_viewer = div
    }
    main_viewer.innerHTML = html_string
  }

  function render_main_view() {
    render_html(`
        <div class="empty-container">
          <h2 class="empty-title">Select an item to read</h2>
          <p class="empty-hint">← Choose a story or feed from the sidebar</p>
        </div>
      `)
  }

  function render_article(data) {
    render_html(article_viewer(data))
  }

  function render_write_page_view(folder_name) {
    render_html(write_page())
  }

  function handle_publish(data) {
    const new_story = {
      title: data.title,
      content: data.content,
      author: 'You',
      date: new Date().toLocaleDateString(),
      description: data.content.slice(0, 100) + '...',
      tags: ['#local', '#' + data.blog.replace(/\s+/g, '')],
      color: '#6366f1'
    }
    save_local_story(new_story)
    alert('Story published locally!')
    render_folder_content('my-stories')
  }

  async function onbatch(batch) {
    console.log('[Wrapper] onbatch:', batch)
    for (const { type, paths } of batch) {
      console.log('[Wrapper] Processing batch type:', type, 'paths:', paths)
      const data = await Promise.all(paths.map(fetch_path_raw_data))
      const valid_data = data.filter(d => d !== null)
      if (valid_data.length > 0) {
        if (type === 'theme') inject(valid_data)
        if (type === 'entries') on_entries(valid_data)
      }
    }

    function fetch_path_raw_data(path) {
      return drive.get(path).then(extract_raw_if_exists)
    }

    function extract_raw_if_exists(file) {
      return file ? file.raw : null
    }
  }

  function inject(data) {
    if (Array.isArray(data)) {
      sheet.replaceSync(data.join('\n'))
    }
  }

  function on_entries(data) {
    console.log('[Wrapper] on_entries called with:', data)
    if (!data || !data[0]) {
      db = graphdb({})
      notify_db_initialized({})
      return
    }
    let parsed_data = {}
    try {
      if (typeof data[0] === 'string') {
        parsed_data = content_parser(data[0]) || {}
      } else {
        parsed_data = data[0]
      }
    } catch (e) { }
    db = graphdb(parsed_data)
    notify_db_initialized(parsed_data)
  }

  function notify_db_initialized(entries) {
    if (send_to_graph_explorer) {
      const head = [by, 'graph_explorer', mid++]
      send_to_graph_explorer({ head, type: 'db_initialized', data: { entries } })
    }
  }

  function graph_explorer_protocol(send) {
    send_to_graph_explorer = send
    return on_graph_explorer_message

    function on_graph_explorer_message(msg) {
      const { type, data } = msg
      if (type === 'selection_changed') {
        const { selected } = data
        if (selected && selected.length > 0) render_folder_content(selected[0])
      }
      if (type.startsWith('db_')) handle_db_request(msg, send)
    }
  }

  async function render_folder_content(path) {
    let drive_path = path
    if (drive_path.startsWith('/')) drive_path = drive_path.slice(1)
    const parts = drive_path.split('/')
    if (parts.length > 2) drive_path = `${parts[0]}/${parts[parts.length - 1]}`

    const data = await try_get_folder_data(drive_path)

    if (data && data.content) {
      render_article(data)
    } else {
      const { folder_name, subs } = get_folder_details(path)

      if (subs.length > 0) {
        const handled = await process_folder_items({ path, subs, folder_name })
        if (handled) return
      }

      render_empty_folder(folder_name, path)
    }
  }

  async function try_get_folder_data(drive_path) {
    try {
      const file = await drive.get(drive_path)
      return content_parser(file.raw)
    } catch (e) {
      return null
    }
  }

  function get_folder_details(path) {
    let folder_name = path.split('/').pop() || path
    let subs = []
    let db_path = path
    if (typeof path === 'string' && path.includes('|')) {
      db_path = path.split('|').pop()
    }

    if (db) {
      const entry = db.get(db_path) || db.get(path)
      if (entry) {
        if (entry.name) folder_name = entry.name
        if (entry.subs) subs = entry.subs
      }
    }
    return { folder_name, subs }
  }

  async function process_folder_items({ path, subs, folder_name }) {
    const fetched_items = (await Promise.all(subs.map(fetch_sub_item))).filter(item => item && item.data)

    async function fetch_sub_item(sub_path) {
      let item_data
      try {
        let sub_drive_path = sub_path
        if (sub_drive_path.startsWith('/')) sub_drive_path = sub_drive_path.slice(1)
        const sub_parts = sub_drive_path.split('/')
        if (sub_parts.length > 2) {
          sub_drive_path = `${sub_parts[0]}/${sub_parts[sub_parts.length - 1]}`
        }
        const file = await drive.get(sub_drive_path)
        item_data = content_parser(file.raw)
      } catch (e) { }
      if (!item_data) return null
      return { path: sub_path, data: item_data }
    }

    if (path.includes('my-stories')) {
      const local_stories = get_local_stories()
      local_stories.forEach(push_local_story)

      function push_local_story(story) {
        fetched_items.push({ path: 'local-' + Date.now() + Math.random(), data: story })
      }
    }

    list_items_data = fetched_items

    if (fetched_items.length > 0) {
      const is_my_stories = path.includes('my-stories') || path.includes('My Stories')

      if (!shadow.adoptedStyleSheets.includes(news_card_sheet)) {
        drive.get('theme/news-card.css').then(apply_news_card_css).catch(ignore_css_error)

        function apply_news_card_css(css_file) {
          news_card_sheet.replaceSync(css_file.raw)
          shadow.adoptedStyleSheets = [...shadow.adoptedStyleSheets, news_card_sheet]
        }

        function ignore_css_error() { }
      }

      const list_html = news_list({ items: fetched_items, folder_name, is_my_stories })

      render_html(list_html)

      const processed_cards = shadow.querySelectorAll('.news-card')
      processed_cards.forEach(map_card_to_path)

      function map_card_to_path(card_el, index) {
        if (fetched_items[index]) {
          card_map.set(card_el, fetched_items[index].path)
        }
      }

      return true
    }
    return false
  }

  function render_empty_folder(folder_name, path) {
    render_html(`
      <div class="empty-folder-container">
        <div class="empty-folder-icon">📂</div>
        <h2 class="empty-folder-title">${folder_name}</h2>
        <p>Select a file inside to view content.</p>
        ${(path.includes('my-stories') || folder_name.includes('My Stories')) ? `<div class="news-fab" data-folder="${folder_name}">+</div>` : ''}
      </div>
    `)
  }
  function get_local_stories() {
    try {
      const stories = localStorage.getItem('p2p_stories')
      return stories ? JSON.parse(stories) : []
    } catch (e) { return [] }
  }

  function save_local_story(story) {
    const stories = get_local_stories()
    stories.unshift(story)
    localStorage.setItem('p2p_stories', JSON.stringify(stories))
  }

  async function handle_db_request(request_msg, send) {
    const { head: request_head, type: operation, data: params } = request_msg
    let result
    if (!db) { send_response(request_head, null); return }

    if (operation === 'db_get') result = db.get(params.path)
    else if (operation === 'db_has') result = db.has(params.path)
    else if (operation === 'db_is_empty') result = db.is_empty()
    else if (operation === 'db_root') result = db.root()
    else if (operation === 'db_keys') result = db.keys()
    else if (operation === 'db_raw') result = db.raw()
    else result = null

    send_response(request_head, result)

    function send_response(request_head, result) {
      const response_head = [by, 'graph_explorer', mid++]
      send({ head: response_head, refs: { cause: request_head }, type: 'db_response', data: { result } })
    }
  }
}

function fallback_module() {
  return {
    _: {
      'graph-explorer': { $: '' },
      './graphdb': { $: '' },
      'article-viewer': { $: '' },
      'news-card': { $: '' },
      'write-page': { $: '' },
      'news-list': { $: '' }
    },
    api: fallback_instance
  }

  function fallback_instance() {
    return {
      _: {
        'graph-explorer': {
          0: override_theme,
          mapping: {
            style: 'theme',
            entries: 'entries',
            runtime: 'runtime',
            mode: 'mode',
            flags: 'flags',
            keybinds: 'keybinds',
            undo: 'undo'
          }
        },
        './graphdb': { 0: '' },
        'article-viewer': { 0: '' },
        'news-card': { 0: '' },
        'write-page': { 0: '' },
        'news-list': { 0: '' }
      },
      drive: {
        'entries/': { 'entries.json': { $ref: './entries.json' } },
        'theme/': {
          'layout.css': {
            $ref: './layout.css'
          },
          'news-card.css': {
            $ref: '../news-card/news-card.css'
          }
        },
        'runtime/': {
          'node_height.json': { raw: '32' },
          'vertical_scroll_value.json': { raw: '0' },
          'horizontal_scroll_value.json': { raw: '0' },
          'selected_instance_paths.json': { raw: '[]' },
          'confirmed_selected.json': { raw: '[]' },
          'instance_states.json': { raw: '{}' },
          'search_entry_states.json': { raw: '{}' },
          'last_clicked_node.json': { raw: 'null' },
          'view_order_tracking.json': { raw: '{}' }
        },
        'flags/': {
          'hubs.json': { raw: '"default"' },
          'selection.json': { raw: 'true' },
          'recursive_collapse.json': { raw: 'true' }
        },
        'keybinds/': { 'navigation.json': { raw: '{}' } },
        'undo/': { 'stack.json': { raw: '[]' } },
        'my-stories/': {
          'story-1': {
            $ref: './data/story-1.md'
          },
          'story-2': {
            $ref: './data/story-2.md'
          },
          'story-3': {
            $ref: './data/story-3.md'
          },
          'story-4': {
            $ref: './data/story-4.md'
          }
        },
        'feeds/': {},
        'feeds/hackers-digest/': {
          'code-coffee': {
            $ref: './data/code-coffee.md'
          },
          'system-design': {
            $ref: './data/system-design.md'
          }
        },
        'feeds/off-the-grid/': {
          'mesh-network': {
            $ref: './data/mesh-network.md'
          },
          fediverse: {
            $ref: './data/fediverse.md'
          },
          'self-hosting': {
            $ref: './data/self-hosting.md'
          }
        },
        'feeds/peer-review/': {
          'network-notes': {
            $ref: './data/network-notes.md'
          }
        },
        'feeds/peer-review/security-chronicles/': {
          'privacy-matters': {
            $ref: './data/privacy-matters.md'
          },
          'zero-trust': {
            $ref: './data/zero-trust.md'
          }
        },
        'lists/': {
          'best-of-tech': {
            $ref: './data/best-of-tech.md'
          },
          'morning-read': {
            $ref: './data/morning-read.md'
          }
        },
        'discover/': {},
        'discover/random-peer-99/': {},
        'discover/satoshi-fan/': {},
        'discover/rust-evangelist/': {}
      }
    }
  }

  function override_theme() {
    return {
      _: {
        mapping: {
          style: 'theme',
          runtime: 'runtime',
          mode: 'mode',
          flags: 'flags',
          keybinds: 'keybinds',
          undo: 'undo',
          entries: 'entries'
        }
      },
      drive: {
        'style/': {
          'style.css': {
            $ref: './style.css'
          }
        },
        'runtime/': {
          'node_height.json': { raw: '32' },
          'vertical_scroll_value.json': { raw: '0' },
          'horizontal_scroll_value.json': { raw: '0' },
          'selected_instance_paths.json': { raw: '[]' },
          'confirmed_selected.json': { raw: '[]' },
          'instance_states.json': {
            raw: JSON.stringify({
              '|/': { expanded_subs: true },
              '|/my-stories': { expanded_subs: true },
              '|/feeds': { expanded_subs: true },
              '|/feeds/hackers-digest': { expanded_subs: true },
              '|/feeds/peer-review': { expanded_subs: true },
              '|/feeds/peer-review/security-chronicles': { expanded_subs: true },
              '|/lists': { expanded_subs: true },
              '|/discover': { expanded_subs: true }
            })
          },
          'search_entry_states.json': { raw: '{}' },
          'last_clicked_node.json': { raw: 'null' },
          'view_order_tracking.json': { raw: '{}' }
        },
        'mode/': {
          'current_mode.json': { raw: '"menubar"' },
          'previous_mode.json': { raw: '"menubar"' },
          'search_query.json': { raw: '""' },
          'multi_select_enabled.json': { raw: 'false' },
          'select_between_enabled.json': { raw: 'false' }
        },
        'flags/': {
          'hubs.json': { raw: '"default"' },
          'selection.json': { raw: 'true' },
          'recursive_collapse.json': { raw: 'true' }
        },
        'keybinds/': {
          'navigation.json': {
            raw: JSON.stringify({
              ArrowUp: 'navigate_up_current_node',
              ArrowDown: 'navigate_down_current_node',
              'Control+ArrowDown': 'toggle_subs_for_current_node',
              'Control+ArrowUp': 'toggle_hubs_for_current_node',
              'Alt+s': 'multiselect_current_node',
              'Alt+b': 'select_between_current_node',
              'Control+m': 'toggle_search_mode',
              'Alt+j': 'jump_to_next_duplicate'
            })
          }
        },
        'undo/': {
          'stack.json': { raw: '[]' }
        }
      }
    }
  }
}
