// news/wrapper.js
const STATE = require('STATE')
console.log('[DEBUG] wrapper.js running, filename:', __filename)
const statedb = STATE(__filename)

const { get } = statedb(fallback_module)
const graph_explorer = require('graph-explorer')
const graphdb = require('./graphdb')

module.exports = my_component_with_graph

async function my_component_with_graph(opts, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const by = id
  let db = null
  let send_to_graph_explorer = null
  let mid = 0

  const on = {
    theme: inject,
    entries: on_entries
  }

  const el = document.createElement('div')
  el.style.height = '100%'
  el.style.width = '100%'
  const shadow = el.attachShadow({ mode: 'closed' })
  const sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [sheet]

  const subs = await sdb.watch(onbatch)
  const explorer_el = await graph_explorer(subs[0], graph_explorer_protocol)
  shadow.append(explorer_el)

  return el

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file ? file.raw : null)))
      const valid_data = data.filter(d => d !== null)
      if (valid_data.length > 0) {
        on[type] && on[type](valid_data)
      }
    }
  }

  function inject(data) {
    sheet.replaceSync(data.join('\n'))
  }



  function on_entries(data) {
    if (!data || !data[0]) {
      // console.error('Entries data is missing or empty.')
      db = graphdb({})
      notify_db_initialized({})
      return
    }

    let parsed_data
    try {
      parsed_data = typeof data[0] === 'string' ? JSON.parse(data[0]) : data[0]
    } catch (e) {
      console.error('Failed to parse entries data:', e)
      parsed_data = {}
    }

    if (typeof parsed_data !== 'object' || !parsed_data) {
      console.error('Parsed entries data is not a valid object.')
      parsed_data = {}
    }

    db = graphdb(parsed_data)
    notify_db_initialized(parsed_data)
  }

  function notify_db_initialized(entries) {
    if (send_to_graph_explorer) {
      const head = [by, 'graph_explorer', mid++]
      send_to_graph_explorer({
        head,
        type: 'db_initialized',
        data: { entries }
      })
    }
  }

  function graph_explorer_protocol(send) {
    send_to_graph_explorer = send
    return on_graph_explorer_message

    function on_graph_explorer_message(msg) {
      const { type } = msg
      if (type.startsWith('db_')) {
        handle_db_request(msg, send)
      }
    }

    async function handle_db_request(request_msg, send) {
      const { head: request_head, type: operation, data: params } = request_msg
      let result

      if (!db) {
        // console.error('[my_component] Database not initialized yet')
        send_response(request_head, null)
        return
      }

      if (operation === 'db_get') {
        result = db.get(params.path)
        if (!result) {
          // Fallback: Try to get from sdb drive directly (for runtime/, mode/ etc)
          // Only try if it looks like a config file path (heuristic)
          // and NOT an internal instance path (|) or entry path (/)
          if (typeof params.path === 'string' && !params.path.startsWith('/') && !params.path.startsWith('|')) {
            console.log('Attempting drive.get for:', params.path)
            try {
              const file = await drive.get(params.path)
              if (file) result = file
            } catch (err) {
              console.error(`Error fetching "${params.path}" from drive:`, err)
            }
          }
        }
      } else if (operation === 'db_has') {
        result = db.has(params.path)
      } else if (operation === 'db_is_empty') {
        result = db.is_empty()
      } else if (operation === 'db_root') {
        result = db.root()
      } else if (operation === 'db_keys') {
        result = db.keys()
      } else if (operation === 'db_raw') {
        result = db.raw()
      } else {
        console.warn('[my_component] Unknown db operation:', operation)
        result = null
      }

      send_response(request_head, result)

      function send_response(request_head, result) {
        const response_head = [by, 'graph_explorer', mid++]
        send({
          head: response_head,
          refs: { cause: request_head },
          type: 'db_response',
          data: { result }
        })
      }
    }
  }
}

function fallback_module() {
  return {
    _: {
      'graph-explorer': { $: '' },
      './graphdb': { $: '' }
    },
    api: fallback_instance
  }

  function fallback_instance() {
    return {
      _: {
        'graph-explorer': {
          0: override_theme,
          mapping: {
            style: 'theme',
            runtime: 'runtime',
            mode: 'mode',
            flags: 'flags',
            keybinds: 'keybinds',
            undo: 'undo'
          }
        },
        './graphdb': {
          0: ''
        }
      },
      drive: {
        'theme/': {
          'style.css': {
            raw: `
              :host {
                display: block;
                height: 100%;
                width: 100%;
              }
              .graph-container {
                color: #abb2bf;
                background-color: #282c34;
                padding: 10px;
                height: 100vh;
                overflow: auto;
              }
              .node {
                display: flex;
                align-items: center;
                white-space: nowrap;
                cursor: default;
                height: 22px;
              }
              .clickable {
                cursor: pointer;
              }
              .node.type-folder > .icon::before { content: 'ðŸ“'; }
              .node.type-js-file > .icon::before { content: 'ðŸ“œ'; }
              .node.type-file > .icon::before { content: 'ðŸ“„'; }
            `
          }
        },
        'entries/': {
          'entries.json': {
            $ref: 'entries.json'
          }
        },
        'runtime/': {},
        'mode/': {},
        'flags/': {},
        'keybinds/': {},
        'undo/': {}
      }
    }
  }
  function override_theme() {
    return {
      _: {},
      drive: {
        'style/': {
          'style.css': {
            raw: `
            :host {
              --primary: #6366f1;
              --bg-sidebar: #f9fafb;
              --border-color: #e5e7eb;
              --text-main: #1f2937;
              --text-muted: #6b7280;

              display: block;
              height: 100%;
              width: 100%;
              font-family: 'Inter', -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif;
              color: var(--text-main);
              background-color: var(--bg-sidebar);
            }

            .graph-container {
              /* aside */
              width: 100%;
              height: 100vh;
              background: var(--bg-sidebar);
              border-right: 1px solid var(--border-color);
              padding: 1rem 0;
              display: flex;
              flex-direction: column;
              overflow: auto;
              box-sizing: border-box;
            }

            .node {
              /* .tree-item */
              display: flex;
              align-items: center;
              gap: 0.5rem;
              padding: 0.4rem 0.75rem; /* slightly more horiz padding */
              margin: 0.1rem 0.5rem;   /* spacing between items */
              border-radius: 6px;
              cursor: pointer;
              font-size: 0.9rem;
              color: var(--text-main);
              transition: background 0.15s;
              height: 32px; /* Fixed height for consistency */
              box-sizing: border-box;
              user-select: none;
            }

            .clickable {
              cursor: pointer;
            }

            .node:hover {
              /* .tree-item:hover */
              background: rgba(0, 0, 0, 0.04);
            }

            /* Map .sidebar-item.active styles */
            .node.selected {
              background: rgba(99, 102, 241, 0.1);
              color: var(--primary);
              font-weight: 600;
            }

            /* HIDE THE MAGIC WAND */
            .wand {
              display: none !important;
            }
            /* Root node hidden for flat sidebar look */
            .node.type-root { display: none !important; }

            /* .folder-icon */
            .icon {
              color: var(--text-muted);
              width: 16px;
              height: 16px;
              display: inline-flex;
              align-items: center;
              justify-content: center;
              margin-right: 0; 
              opacity: 1;      
            }

            /* Expansion/Collapse Trigger (Prefix) */
            .prefix {
              cursor: pointer;
              min-width: 16px; 
              height: 100%;
              display: inline-flex;
              align-items: center;
              justify-content: center;
            }

            .indent {
              height: 100%;
              display: inline-flex;
            }
            
            /* Indentation Steps */
            .pipe, .blank {
              display: inline-block;
              width: 24px; /* Width per indentation level */
              height: 100%;
            }
            
            /* Visual indicator for expansion if missing */
            .prefix::after {
              content: 'â–¸';
              font-size: 10px;
              color: var(--text-muted);
              display: inline-block;
              transition: transform 0.15s;
            }
            .prefix.expanded::after {
               transform: rotate(90deg);
            }
            /* Hide prefix content if it was something else, and rely on our after */
            /* This is a guess, but if graph-explorer leaves it empty, we are good. */

            /* Custom Icons for Main Sections */
            .node.type-stories > .icon::before { content: 'ðŸ“'; font-size: 14px; }
            .node.type-feeds > .icon::before   { content: 'ðŸ“¡'; font-size: 14px; }
            .node.type-lists > .icon::before   { content: 'ðŸ“‹'; font-size: 14px; }
            .node.type-discover > .icon::before { content: 'ðŸ§­'; font-size: 14px; }

            /* Specific Icons - Simplifying to Standard Sidebar look */
            /* Generic Folders */
            .node.type-folder > .icon { 
              display: none;
            }
            
            /* Files - Removing "Word" icon look, making it generic */
            .node.type-file > .icon::before,
            .node.type-js-file > .icon::before { 
              content: ''; /* Remove emoji */
              width: 6px;
              height: 6px;
              background-color: var(--text-muted);
              border-radius: 50%; /* Simple dot for files */
              display: block;
            }
            
            /* Selected State Icon Color Override */
            .node.selected .icon {
              color: var(--primary);
            }
            .node.selected.type-file > .icon::before,
            .node.selected.type-js-file > .icon::before {
              background-color: var(--primary);
            }
            
            /* Clean Scrollbars */
            .graph-container::-webkit-scrollbar {
              width: 6px;
            }
            .graph-container::-webkit-scrollbar-track {
              background: transparent;
            }
            .graph-container::-webkit-scrollbar-thumb {
              background: #d1d5db;
              border-radius: 3px;
            }
            .graph-container::-webkit-scrollbar-thumb:hover {
              background: #9ca3af;
            }
          `
          }
        },
        'runtime/': {
          'node_height.json': { raw: '16' },
          'vertical_scroll_value.json': { raw: '0' },
          'horizontal_scroll_value.json': { raw: '0' },
          'selected_instance_paths.json': { raw: '[]' },
          'confirmed_selected.json': { raw: '[]' },
          'instance_states.json': {
            raw: JSON.stringify({
              '|/': { expanded_subs: true },
              '|/my-stories': { expanded_subs: true },
              '|/feeds': { expanded_subs: true },
              '|/feeds/hackers-digest': { expanded_subs: true },
              '|/feeds/peer-review': { expanded_subs: true },
              '|/feeds/peer-review/security-chronicles': { expanded_subs: true },
              '|/lists': { expanded_subs: true },
              '|/discover': { expanded_subs: true }
            })
          },
          'search_entry_states.json': { raw: '{}' },
          'last_clicked_node.json': { raw: 'null' },
          'view_order_tracking.json': { raw: '{}' }
        },
        'mode/': {
          'current_mode.json': { raw: '"menubar"' },
          'previous_mode.json': { raw: '"menubar"' },
          'search_query.json': { raw: '""' },
          'multi_select_enabled.json': { raw: 'false' },
          'select_between_enabled.json': { raw: 'false' }
        },
        'flags/': {
          'hubs.json': { raw: '"default"' },
          'selection.json': { raw: 'false' },
          'recursive_collapse.json': { raw: 'true' }
        },
        'keybinds/': {
          'navigation.json': {
            raw: JSON.stringify({
              ArrowUp: 'navigate_up_current_node',
              ArrowDown: 'navigate_down_current_node',
              'Control+ArrowDown': 'toggle_subs_for_current_node',
              'Control+ArrowUp': 'toggle_hubs_for_current_node',
              'Alt+s': 'multiselect_current_node',
              'Alt+b': 'select_between_current_node',
              'Control+m': 'toggle_search_mode',
              'Alt+j': 'jump_to_next_duplicate'
            })
          }
        },
        'undo/': {
          'stack.json': { raw: '[]' }
        }
      }
    }
  }
}
