// newsfeed_card_list/index.js
const STATE = require('STATE')
const statedb = STATE(__filename)

const { get } = statedb(fallback_module)
const news_cards = require('news_cards')

module.exports = newsfeed_card_list

async function newsfeed_card_list (opts, protocol) {
  const { sid } = opts
  const { id, sdb } = await get(sid)
  const { drive } = sdb

  const viewer_file = await drive.get('runtime/viewer_data.json')
  const viewer_data = JSON.parse(viewer_file ? viewer_file.raw : '{}')
  const data = viewer_data.data || {}
  const items = data.items || []

  const max_items = items.slice(0, 30)

  const subs = await sdb.watch(onbatch)
  const card_subs = (subs || []).sort(sort_by_index)

  const items_html = await Promise.all(max_items.map(render_card))

  return `
    <div class="newsfeed-list">
      ${items_html.join('')}
    </div>
  `

  async function render_card (item, index) {
    if (index >= card_subs.length) return ''
    const card_sub = card_subs[index]

    await sdb.drive.put(`runtime/card-${index}.json`, JSON.stringify({ data: item.data }))

    return news_cards({ sid: card_sub.sid })
  }

  function sort_by_index (a, b) {
    return a.sid[a.sid.length - 1] - b.sid[b.sid.length - 1]
  }

  function onbatch () {}
}

function fallback_module () {
  const _ = {
    news_cards: { $: '' }
  }

  return { _, api: fallback_instance }

  function fallback_instance () {
    const _ = {
      news_cards: {
        mapping: { runtime: 'runtime', theme: 'theme' }
      }
    }

    for (let i = 0; i < 30; i++) {
        _['news_cards'][i] = ''
    }

    const drive = {
      'runtime/': {
        'viewer_data.json': { raw: '{}' }
      }
    }

    return { _, drive }
  }
}
