// newsfeed_card_list/index.js
const STATE = require('STATE')
const statedb = STATE(__filename)

const { get } = statedb(fallback_module)
const news_cards = require('news_cards')

module.exports = newsfeed_card_list

async function newsfeed_card_list (opts, protocol) {
  const { sid } = opts
  const { id, sdb } = await get(sid)
  const { drive } = sdb

  const viewer_file = await drive.get('runtime/viewer_data.json')
  const viewer_data = JSON.parse(viewer_file ? viewer_file.raw : '{}')
  const items = viewer_data.data?.items || []

  const cards_html = await Promise.all(items.map(render_card))

  return `
    <div class="newsfeed-card-list">
      ${cards_html.join('')}
    </div>
  `

  async function render_card (item, index) {
    const card_sid = [...id, 'news_cards', index]

    await sdb.drive.put(`runtime/card-${index}.json`, JSON.stringify({ data: item.data }))

    return news_cards({
      sid: card_sid,
      index
    })
  }
}

function fallback_module () {
  return {
    drive: {
      'runtime/': {
        'viewer_data.json': { raw: '{}' }
      }
    },
    _: {
      news_cards: {
        $: '',
        mapping: { runtime: 'runtime', theme: 'theme' }
      }
    },
    api: fallback_instance
  }

  function fallback_instance () {
    return {
      drive: {
        'runtime/': {
          'viewer_data.json': { raw: '{}' }
        }
      },
      _: {
        news_cards: {
          0: '',
          '*': '',
          mapping: { runtime: 'runtime', theme: 'theme' }
        }
      }
    }
  }
}
