// newsfeed_card_list/index.js
const STATE = require('STATE')
const statedb = STATE(__filename)

const { get } = statedb(fallback_module)
const news_cards = require('news_cards')

module.exports = newsfeed_card_list

async function newsfeed_card_list (opts, protocol) {
  const { sid } = opts
  const { sdb } = await get(sid)

  const el = document.createElement('div')
  el.className = 'newsfeed-list'

  const card_sends = []
  const card_els = []
  let send_to_parent

  function handle_message (msg) {
    if (msg.type === 'provision') update_cards(msg.data)
  }

  if (protocol) send_to_parent = protocol(handle_message)

  const subs = await sdb.watch(onbatch)
  const card_subs = (subs || []).sort(sort_by_index)

  for (let i = 0; i < card_subs.length; i++) {
    const card_el = await news_cards({ sid: card_subs[i].sid }, make_card_protocol(i))
    card_els.push(card_el)
  }

  return el

  function make_card_protocol (idx) {
    return function card_protocol (child_handler) {
      card_sends[idx] = child_handler
      return handle_card_message
    }
  }

  function handle_card_message (msg) {
    if (msg.type === 'card_clicked' && send_to_parent) {
      send_to_parent({ type: 'card_clicked', data: msg.data })
    }
  }

  function update_cards (payload) {
    const items = payload.items || []

    while (el.firstChild) el.removeChild(el.firstChild)

    for (let i = 0; i < items.length && i < card_els.length; i++) {
      if (card_sends[i]) {
        card_sends[i]({ type: 'provision', data: { data: items[i].data, path: items[i].path } })
      }
      el.appendChild(card_els[i])
    }
  }

  function sort_by_index (a, b) {
    return a.sid[a.sid.length - 1] - b.sid[b.sid.length - 1]
  }

  function onbatch () {}
}

function fallback_module () {
  const _ = {
    news_cards: { $: '' }
  }

  return { _, api: fallback_instance }

  function fallback_instance () {
    const _ = {
      news_cards: {
        mapping: { runtime: 'runtime', theme: 'theme' }
      }
    }

    for (let i = 0; i < 30; i++) {
      _.news_cards[i] = ''
    }

    const drive = {
      'runtime/': {
        'viewer_data.json': { raw: '{}' }
      }
    }

    return { _, drive }
  }
}
