// identity-networking/index.js
// Identity Networking - creates swarm + replicates store on connection
// Also provides create_app_network for apps to get a vault-filtered swarm
// Handles identity protocol exchange + WebRTC upgrade for browser peers
const b4a = require('b4a')
const Hyperswarm = require('hyperswarm')
const Hypercore = require('hypercore')
const DHT = require('@hyperswarm/dht-relay')
const Stream = require('@hyperswarm/dht-relay/ws')
const Protomux = require('protomux')
const c = require('compact-encoding')
const HyperWebRTC = require('hyper-webrtc')

module.exports = { network, create_app_network }

const is_cli = typeof globalThis.open === 'undefined'

/***************************************
IDENTITY NETWORK — own devices only
***************************************/

// store for replication
// mnemonic data to join the swarm with a determinstic keypair
async function network (store, mnemonic_data) {
  await store.ready()
  const swarm = await create_swarm(mnemonic_data)
  console.log('[identity] swarm joined')
  // for now its replicating the entire store, but this connection is only between our own devices
  swarm.on('connection', function (socket) {
    console.log('[identity] new connection')
    upgrade_connection(socket, function (connection) { store.replicate(connection)  })})
  return { swarm }
}

/***************************************
APP NETWORK — vault-filtered swarm for apps
***************************************/

async function create_app_network (store) {
  const is_vault_core = store._is_vault_core
  const swarm = await create_swarm()
  const peer_streams = []

  store.on('core-open', function (core) {
    if (is_vault_core(b4a.toString(core.discoveryKey, 'hex'))) return
    for (const stream of peer_streams) core.replicate(stream, { session: true })
  })

  // Now that we have two swarms, we will upgrade the connection to webrtc for identity swarm as well
  swarm.on('connection', function (socket) {
    if (!socket.userData) socket.userData = null
    upgrade_connection(socket, function (connection) {
      replicate(connection)
    })
  })

  function replicate (connection) {
    const stream = Hypercore.createProtocolStream(connection, { ondiscoverykey })
    peer_streams.push(stream)
    stream.once('close', function () {
      const idx = peer_streams.indexOf(stream)
      if (idx !== -1) peer_streams.splice(idx, 1)
    })
    return stream

    function ondiscoverykey (discovery_key) {
      const id = b4a.toString(discovery_key, 'hex')
      const core = store.cores.get(id)
      if (!core) return
      if (is_vault_core(id)) return
      core.replicate(stream, { session: true })
    }
  }

  return { swarm, replicate }
}

/***************************************
UPGRADE CONNECTION
identity exchange + Webrtc upgrade
both peers exchange mode (browser/native).
browser-to-browser: upgrade to WebRTC
browser-native or native-native raw socket
***************************************/
function upgrade_connection (socket, on_ready) {
  const mux = Protomux.from(socket)
  const mode = is_cli ? 'native' : 'browser'
  const channel = mux.createChannel({ protocol: 'hyper-webrtc/upgrade', onopen })
  channel.open()

  function onopen () {
    const message = channel.addMessage({ encoding: c.json, onmessage })
    message.send({ type: 'identity', data: { mode } })

    function onmessage (msg) {
      const peer_mode = msg.data?.mode
      if (is_cli || peer_mode === 'native') return on_ready(socket)
      const rtc = HyperWebRTC.from(socket, { initiator: socket.isInitiator })
      console.log('[p2p-networking] WebRTC connection established.')
      rtc.on('open', function () { on_ready(rtc) })
    }
  }
}

/***************************************
SWARM CREATION
***************************************/

async function create_swarm (mnemonic_data) {
  if (is_cli) {
    return new Hyperswarm(mnemonic_data ? { keyPair: mnemonic_data.keypair } : {})
  }
  const is_dev = location.hostname === 'localhost' || location.hostname.startsWith('192.') || location.hostname.startsWith('10.')
  const relay_url = is_dev ? 'ws://localhost:8080' : 'wss://relay-production-9c0e.up.railway.app'
  return new Promise(function (resolve, reject) {
    const ws = new WebSocket(relay_url)
    ws.addEventListener('error', reject)
    ws.addEventListener('open', function () {
      const opts = mnemonic_data
        ? { dht: new DHT(new Stream(true, ws)), keyPair: mnemonic_data.keypair }
        : { dht: new DHT(new Stream(true, ws)) }
      resolve(new Hyperswarm(opts))
    })
  })
}
