diff --git a/node_modules/bip39-mnemonic/index.js b/node_modules/bip39-mnemonic/index.js
index 541e5e9..4c1b999 100644
--- a/node_modules/bip39-mnemonic/index.js
+++ b/node_modules/bip39-mnemonic/index.js
@@ -1,4 +1,4 @@
-const sodium = require('sodium-universal')
+const sodium = require('../../src/node_modules/sodium')
 const b4a = require('b4a')
 const assert = require('nanoassert')
 const { detectLanguage, loadWordlist } = require('./wordlist')
