// newsfeed_view/index.js
const STATE = require('STATE')
const statedb = STATE(__filename)

const { get } = statedb(fallback_module)
const news_list = require('newsfeed_card_list')
const shopts = { mode: 'closed' }

module.exports = newsfeed_view

async function newsfeed_view (opts, protocol) {
  const { sid } = opts
  const { id, sdb } = await get(sid)
  const { drive } = sdb

  let send
  if (protocol) send = protocol(handle_message)

  const element = document.createElement('div')
  const shadow = element.attachShadow(shopts)

  const layout_sheet = new CSSStyleSheet()
  const card_sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [layout_sheet, card_sheet]

  const paths = ['runtime/viewer_data.json', 'theme/layout.css', 'theme/news-card.css']
  const files = await Promise.all(paths.map(path => drive ? drive.get(path).catch(() => null) : null))
  const [viewer_file, layout_css, newscard_css] = files

  if (layout_css && layout_css.raw) layout_sheet.replaceSync(layout_css.raw)
  if (newscard_css && newscard_css.raw) card_sheet.replaceSync(newscard_css.raw)

  const viewer_data = JSON.parse(viewer_file && viewer_file.raw ? viewer_file.raw : '{}')
  const data = viewer_data.data || {}

  const subs = await sdb.watch(handle_watch)

  if (data.items) {
    const list_sub = subs && subs.length > 0 ? subs[0] : null
    if (list_sub) {
      const list_html = await news_list({
        sid: list_sub.sid
      })
      shadow.innerHTML = list_html
    }
  } else if (data.content) {
    const content_html = data.content
      .split('\n\n')
      .map(parse_block)
      .join('')

    shadow.innerHTML = `
    <article class="article-container">
      <header class="article-header">
        <h1 class="article-title">${data.title || 'Untitled'}</h1>
        <div class="article-meta">
          <span>By <strong>${data.author || 'Anonymous'}</strong></span> • <span>${data.date || 'Unknown Date'}</span>
        </div>
      </header>
      <div class="article-body">${content_html}</div>
    </article>
    `
  }

  shadow.addEventListener('click', handle_shadow_click)

  return element

  function handle_shadow_click (e) {
    const card_target = e.target.closest('.news-card')
    if (card_target && send && data.items) {
      const path = data.items.find(match_card, { target: card_target })?.path

      if (path) {
        send({ type: 'select_node', data: { instance_path: path } })
      }
    }

    function match_card (item, index) {
      const card_el = shadow.querySelectorAll('.news-card')[index]
      return card_el === this.target
    }
  }

  function handle_watch (batch) {
  }

  function parse_block (block) {
    block = block.trim()
    if (!block) return ''

    if (block.startsWith('# ')) return `<h1>${block.slice(2)}</h1>`
    if (block.startsWith('## ')) return `<h2>${block.slice(3)}</h2>`
    if (block.startsWith('### ')) return `<h3>${block.slice(4)}</h3>`

    if (block.startsWith('- ')) {
      const lines = block.split('\n')
      const items = lines.map(format_list_item).join('')
      return `<ul>${items}</ul>`
    }

    let p = block

    p = p.replace(/\*\*(.*?)\*\*/g, '<strong>$1</strong>')
    p = p.replace(/\*(.*?)\*/g, '<em>$1</em>')
    p = p.replace(/\[(.*?)\]\((.*?)\)/g, '<a href="$2">$1</a>')

    return `<p>${p}</p>`
  }

  function format_list_item (line) {
    return `<li>${line.replace(/^- /, '')}</li>`
  }

  function handle_message (msg) {
  }
}

function fallback_module () {
  const _ = {
    newsfeed_card_list: { $: '' }
  }

  return { _, api: fallback_instance }

  function fallback_instance () {
    const _ = {
      newsfeed_card_list: {
        0: '',
        mapping: { runtime: 'runtime', theme: 'theme' }
      }
    }

    const drive = {
      'runtime/': {
        'viewer_data.json': { raw: '{}' }
      }
    }

    return { _, drive }
  }
}
