// newsfeed_view/index.js
const STATE = require('STATE')
const statedb = STATE(__filename)

const { get } = statedb(fallback_module)
const news_list = require('newsfeed_card_list')

module.exports = newsfeed_view

async function newsfeed_view (opts, protocol) {
  const { sid } = opts
  const { id, sdb } = await get(sid)
  const { drive } = sdb

  let send_to_parent

  function handle_message (msg) {
    if (msg.type === 'render_feed') show_feed(msg.data)
    if (msg.type === 'render_article') show_article(msg.data)
  }

  if (protocol) send_to_parent = protocol(handle_message)

  const element = document.createElement('div')
  const shadow = element.attachShadow({ mode: 'closed' })

  const layout_sheet = new CSSStyleSheet()
  const card_sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [layout_sheet, card_sheet]

  const [layout_css, newscard_css] = await Promise.all([
    drive.get('theme/layout.css').catch(return_null),
    drive.get('theme/news-card.css').catch(return_null)
  ])

  if (layout_css && layout_css.raw) layout_sheet.replaceSync(layout_css.raw)
  if (newscard_css && newscard_css.raw) card_sheet.replaceSync(newscard_css.raw)

  let list_el = null
  let send_to_list = null

  const subs = await sdb.watch(handle_watch)
  const list_sub = subs && subs.length > 0 ? subs[0] : null

  if (list_sub) {
    list_el = await news_list({ sid: list_sub.sid }, list_protocol)
  }

  shadow.addEventListener('click', handle_shadow_click)

  return element

  function list_protocol (child_handler) {
    send_to_list = child_handler
    return handle_list_message
  }

  function handle_list_message (msg) {
    if (msg.type === 'card_clicked' && send_to_parent) {
      send_to_parent({ type: 'select_node', data: { instance_path: msg.data.path } })
    }
  }

  function show_feed (data) {
    clear_shadow()
    if (list_el) shadow.appendChild(list_el)
    if (send_to_list) {
      send_to_list({ type: 'provision', data: data })
    }
  }

  function show_article (payload) {
    clear_shadow()
    const data = payload.data || payload
    const content_html = (data.content || '')
      .split('\n\n')
      .map(parse_block)
      .join('')

    shadow.innerHTML = `
      <article class="news-container">
        <button class="back-btn">← Back to feed</button>
        <header class="news-header">
          <div>
            <h2 class="news-title article-title-main">${data.title || 'Untitled'}</h2>
            <div class="news-subheader">
              <span>By <strong>${data.author || 'Anonymous'}</strong></span> • <span>${data.date || 'Unknown Date'}</span>
            </div>
          </div>
        </header>
        <div class="article-body">${content_html}</div>
      </article>
    `
  }

  function clear_shadow () {
    while (shadow.firstChild) shadow.removeChild(shadow.firstChild)
  }

  function return_null () {
    return null
  }

  function handle_shadow_click (e) {
    if (e.target.closest('.back-btn') && send_to_parent) {
      handle_back_click()
    }
  }

  function handle_back_click () {
    send_to_parent({ type: 'navigate_back' })
  }

  function handle_watch () {}

  function parse_block (block) {
    block = block.trim()
    if (!block) return ''
    if (block.startsWith('# ')) return `<h1>${block.slice(2)}</h1>`
    if (block.startsWith('## ')) return `<h2>${block.slice(3)}</h2>`
    if (block.startsWith('### ')) return `<h3>${block.slice(4)}</h3>`
    if (block.startsWith('- ')) {
      const items = block.split('\n').map(format_list_item).join('')
      return `<ul>${items}</ul>`
    }
    let p = block
    p = p.replace(/\*\*(.*?)\*\*/g, '<strong>$1</strong>')
    p = p.replace(/\*(.*?)\*/g, '<em>$1</em>')
    p = p.replace(/\[(.*?)\]\((.*?)\)/g, '<a href="$2">$1</a>')
    return `<p>${p}</p>`
  }

  function format_list_item (line) {
    return `<li>${line.replace(/^- /, '')}</li>`
  }
}

function fallback_module () {
  const _ = {
    newsfeed_card_list: { $: '' }
  }

  return { _, api: fallback_instance }

  function fallback_instance () {
    const _ = {
      newsfeed_card_list: {
        0: '',
        mapping: { runtime: 'runtime', theme: 'theme' }
      }
    }

    const drive = {
      'runtime/': {
        'viewer_data.json': { raw: '{}' }
      },
      'theme/': {}
    }

    return { _, drive }
  }
}
