// menu_sidebar/index.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)

const graph_explorer = require('graph-explorer')

module.exports = menu_sidebar

async function menu_sidebar (opts, protocol) {
  const { sid } = opts
  const { sdb } = await get(sid)

  let send_to_parent = null
  let send_to_explorer = null

  const el = document.createElement('div')
  el.className = 'sidebar-container'

  if (protocol) {
    send_to_parent = protocol(on_parent_message)
  }

  const subs = await sdb.watch(onbatch)

  if (subs && subs.length > 0) {
    const explorer_el = await graph_explorer({ sid: subs[0].sid }, explorer_protocol)
    el.appendChild(explorer_el)
  }

  return el

  function onbatch (batch) { }

  function explorer_protocol (send) {
    send_to_explorer = send
    return on_explorer_message
  }

  function on_explorer_message (msg) {
    if (send_to_parent) send_to_parent(msg)
  }

  function on_parent_message (msg) {
    if (send_to_explorer) send_to_explorer(msg)
  }
}

function fallback_module () {
  const _ = {
    'graph-explorer': { $: '' }
  }

  return { _, api: fallback_instance }

  function fallback_instance () {
    const _ = {
      'graph-explorer': {
        0: override_theme,
        '*': override_theme,
        mapping: {
          style: 'style',
          entries: 'entries',
          runtime: 'runtime',
          mode: 'mode',
          flags: 'flags',
          keybinds: 'keybinds',
          undo: 'undo'
        }
      }
    }

    const drive = {
      'style/': { 'style.css': { $ref: 'style.css' } },
      'entries/': { 'entries.json': { $ref: '../news/entries.json' } },
      'runtime/': {
        'node_height.json': { raw: '32' },
        'vertical_scroll_value.json': { raw: '0' },
        'horizontal_scroll_value.json': { raw: '0' },
        'selected_instance_paths.json': { raw: '[]' },
        'confirmed_selected.json': { raw: '[]' },
        'instance_states.json': { raw: '{}' },
        'search_entry_states.json': { raw: '{}' },
        'last_clicked_node.json': { raw: 'null' },
        'view_order_tracking.json': { raw: '{}' }
      },
      'mode/': {
        'current_mode.json': { raw: '"menubar"' },
        'previous_mode.json': { raw: '"menubar"' },
        'search_query.json': { raw: '""' },
        'multi_select_enabled.json': { raw: 'false' },
        'select_between_enabled.json': { raw: 'false' }
      },
      'flags/': {
        'hubs.json': { raw: '"default"' },
        'selection.json': { raw: 'true' },
        'recursive_collapse.json': { raw: 'true' }
      },
      'keybinds/': {
        'navigation.json': {
          raw: JSON.stringify({
            ArrowUp: 'navigate_up_current_node',
            ArrowDown: 'navigate_down_current_node',
            'Control+ArrowDown': 'toggle_subs_for_current_node',
            'Control+ArrowUp': 'toggle_hubs_for_current_node',
            'Alt+s': 'multiselect_current_node',
            'Alt+b': 'select_between_current_node',
            'Control+m': 'toggle_search_mode',
            'Alt+j': 'jump_to_next_duplicate'
          })
        }
      },
      'undo/': { 'stack.json': { raw: '[]' } }
    }

    return { _, drive }
  }

  function override_theme () {
    return {
      _: {
        mapping: {
          style: 'style',
          runtime: 'runtime',
          mode: 'mode',
          flags: 'flags',
          keybinds: 'keybinds',
          undo: 'undo',
          entries: 'entries'
        }
      },
      drive: {
        'style/': {
          'style.css': {
            $ref: 'style.css'
          }
        },
        'runtime/': {
          'node_height.json': { raw: '32' },
          'vertical_scroll_value.json': { raw: '0' },
          'horizontal_scroll_value.json': { raw: '0' },
          'selected_instance_paths.json': { raw: '[]' },
          'confirmed_selected.json': { raw: '[]' },
          'instance_states.json': {
            raw: JSON.stringify({
              '|/': { expanded_subs: true },
              '|/my-stories': { expanded_subs: true },
              '|/feeds': { expanded_subs: true },
              '|/feeds/hackers-digest': { expanded_subs: true },
              '|/feeds/peer-review': { expanded_subs: true },
              '|/feeds/peer-review/security-chronicles': { expanded_subs: true },
              '|/lists': { expanded_subs: true },
              '|/discover': { expanded_subs: true }
            })
          },
          'search_entry_states.json': { raw: '{}' },
          'last_clicked_node.json': { raw: 'null' },
          'view_order_tracking.json': { raw: '{}' }
        },
        'mode/': {
          'current_mode.json': { raw: '"menubar"' },
          'previous_mode.json': { raw: '"menubar"' },
          'search_query.json': { raw: '""' },
          'multi_select_enabled.json': { raw: 'false' },
          'select_between_enabled.json': { raw: 'false' }
        },
        'flags/': {
          'hubs.json': { raw: '"default"' },
          'selection.json': { raw: 'true' },
          'recursive_collapse.json': { raw: 'true' }
        },
        'keybinds/': {
          'navigation.json': {
            raw: JSON.stringify({
              ArrowUp: 'navigate_up_current_node',
              ArrowDown: 'navigate_down_current_node',
              'Control+ArrowDown': 'toggle_subs_for_current_node',
              'Control+ArrowUp': 'toggle_hubs_for_current_node',
              'Alt+s': 'multiselect_current_node',
              'Alt+b': 'select_between_current_node',
              'Control+m': 'toggle_search_mode',
              'Alt+j': 'jump_to_next_duplicate'
            })
          }
        },
        'undo/': {
          'stack.json': { raw: '[]' }
        }
      }
    }
  }
}
