// local-storage-shim/index.js
/*
this is basically for the subscription management for cli peers
as the modules and functions expect "localStorage", we will just shim
the node/bare filesystem to pass into our functions..
 */
const fs = require('bare-fs')
const path = require('bare-path')
const process = require('bare-process')

const storage_dir = path.join(process.cwd(), '.local-storage')
const storage_file = path.join(storage_dir, 'storage.json')

// Ensure storage directory exists
try {
  if (!fs.existsSync(storage_dir)) {
    fs.mkdirSync(storage_dir, { recursive: true })
  }
} catch (err) {
  console.error('Failed to create storage directory:', err)
}

// In-memory cache
let storage_cache = {}

// Load storage from file
function load_storage () {
  try {
    if (fs.existsSync(storage_file)) {
      const data = fs.readFileSync(storage_file, 'utf8')
      storage_cache = JSON.parse(data)
    }
  } catch (err) {
    console.error('Failed to load storage:', err)
    storage_cache = {}
  }
}

// Save storage to file
function save_storage () {
  try {
    fs.writeFileSync(storage_file, JSON.stringify(storage_cache, null, 2))
  } catch (err) {
    console.error('Failed to save storage:', err)
  }
}

// Initialize storage
load_storage()

// localStorage API implementation
const localStorage = {
  getItem (key) {
    return storage_cache[key] || null
  },

  setItem (key, value) {
    storage_cache[key] = String(value)
    save_storage()
  },

  removeItem (key) {
    delete storage_cache[key]
    save_storage()
  },

  clear () {
    storage_cache = {}
    save_storage()
  },

  reset_all () {
    storage_cache = {}
    try { fs.rmSync(storage_dir, { recursive: true, force: true }) } catch { }
    const cwd = process.cwd()
    const entries = fs.readdirSync(cwd)
    for (const entry of entries) {
      if (entry.startsWith('storage-') || entry.endsWith('-seedphrase')) {
        fs.rmSync(path.join(cwd, entry), { recursive: true, force: true })
      }
    }
  },

  key (index) {
    const keys = Object.keys(storage_cache)
    return keys[index] || null
  },

  get length () {
    return Object.keys(storage_cache).length
  }
}

// make it global
if (typeof global !== 'undefined' && !global.localStorage) {
  global.localStorage = localStorage
}

module.exports = localStorage
