// write_page/index.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)

module.exports = write_page

async function write_page (opts, protocol) {
  const { sid } = opts
  const { sdb } = await get(sid)
  const { drive } = sdb

  let send
  if (protocol) send = protocol(handle_message)

  const cfg_file = await drive.get('runtime/write_data.json')
  const config = JSON.parse(cfg_file && cfg_file.raw ? cfg_file.raw : '{}')
  const selected_blog = config.selected_blog || 'Main Blog'

  const el = document.createElement('div')
  
  const blogs = ['Main Blog', 'Tech Weekly', 'Cooking Adventures', 'Travel Logs']
  const options = blogs.map(render_blog_option).join('')

  const tip_data = [
    { title: 'Be Authentic', text: 'Write what you genuinely think and feel, not what algorithms demand' },
    { title: 'Tell a Story', text: 'Use examples and narratives to engage readers and make ideas stick' },
    { title: 'Add Value', text: 'Help readers learn something new or see the world differently' }
  ]

  const tips_html = tip_data.map(render_tip).join('')

  el.innerHTML = `
  <div class="write-page-container">
    <div class="section-header">
      <h1>Write a Story</h1>
      <p>Share your thoughts with the network</p>
    </div>

    <div class="card">
      <form class="space-y-8" id="write-story-form">
        <div class="input-group">
          <label>Publishing To</label>
          <select class="blog-select" name="blog">
            ${options}
          </select>
        </div>

        <div class="input-group">
          <label>Story Title</label>
          <input type="text" class="input-title" name="title" placeholder="Give your story a captivating title..." required>
        </div>

        <div class="divider"></div>

        <div class="input-group">
          <label>Your Story</label>
          <textarea class="input-content" name="content" placeholder="Write your story here. Share your thoughts, experiences, and insights..." required></textarea>
          <div class="word-count">
            <span id="word-count-span">0 words</span>
            <span id="read-time-span">~0 min read</span>
          </div>
        </div>

        <div class="actions">
          <button type="submit" class="btn-publish">Publish Story</button>
          <p class="action-text">Your story will be stored locally and synced with your network</p>
        </div>
      </form>
    </div>

    <div class="tips">
      ${tips_html}
    </div>
  </div>
  `

  return el

  function render_blog_option (blog) {
    return `<option value="${blog}"${blog === selected_blog ? ' selected' : ''}>${blog}</option>`
  }

  function render_tip (t) {
    return `
    <div class="tip">
      <h3>${t.title}</h3>
      <p>${t.text}</p>
    </div>
  `
  }

  function handle_message (msg) {}
}

function fallback_module () {
  return { api: fallback_instance }

  function fallback_instance () {
    const drive = {
      'runtime/': {
        'write_data.json': { raw: '{}' }
      }
    }
    return { drive }
  }
}
