// write_page/index.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const net_helper = require('net_helper')

module.exports = write_page

async function write_page (opts, invite) {
  const { sid } = opts
  const { id } = await get(sid)

  const { io } = net_helper(id)
  io.on.news = io_news()
  if (invite) io.accept(invite)

  const el = document.createElement('div')

  render_form('Main Blog')

  return el

  function io_news () {
    const on = { provision: on_provision }
    return protocol
    function protocol (m) { (on[m.type] || on_fail)(m) }
    function on_fail () { }
    function on_provision (m) { update_form(m.data) }
  }

  function render_form (blog_name) {
    el.innerHTML = `
      <div class="write-page-container">
        <div class="section-header">
          <h1>Write a Story</h1>
          <p class="write-subtitle">Create something amazing for <strong>${blog_name}</strong></p>
        </div>

        <div class="card space-y-8">
          <form id="write-story-form">
            <div class="input-group">
              <label>Target Blog</label>
              <input type="text" name="blog" value="${blog_name}" readonly class="blog-select">
            </div>

            <div class="input-group">
              <label>Story Title</label>
              <input type="text" name="title" placeholder="Enter your story title..." class="input-title" required autofocus>
            </div>

            <div class="divider"></div>

            <div class="input-group">
              <label>Content</label>
              <textarea name="content" placeholder="Start writing your story here..." class="input-content" required></textarea>
            </div>

            <div class="word-count">
              <span id="word-count-span">0 words</span>
              <span id="read-time-span">~0 min read</span>
            </div>

            <div class="actions">
              <button type="submit" class="btn-publish">Publish Story</button>
              <span class="action-text">Saved to local draft</span>
            </div>
          </form>
        </div>
      </div>
    `
  }

  function update_form (config) {
    const blog_name = config.selected_blog || 'Main Blog'
    const blog_input = el.querySelector('.blog-select')
    const subtitle = el.querySelector('.write-subtitle')
    if (blog_input) blog_input.value = blog_name
    if (subtitle) subtitle.innerHTML = `Create something amazing for <strong>${blog_name}</strong>`
  }
}

function fallback_module () {
  const _ = {
    net_helper: { $: '' }
  }

  return { _, api: fallback_instance }

  function fallback_instance () {
    const drive = {
      'runtime/': {
        'write_data.json': { raw: '{}' }
      }
    }
    return { drive }
  }
}
