// identity/index.js
// Identity module which is app independent
// This module handles: keypairs, pairing, networking, data structures
// Doesnt contain app-specific things
// Exports a single constructor function that returns {make, load} API

const b4a = require('b4a')
const Corestore = require('corestore')
const pairing_manager_constructor = require('pairing-manager')
const datastructure_manager = require('datastructure-manager')
const { create_autobee } = require('autobee')
const { create_auditcore } = require('auditcore')
const { network, create_app_network } = require('identity-networking')
const { create_vault_ops, make_emitter, parse_vault_invite } = require('identity-operations')
const { get_or_create_keypair } = require('crypto-helpers')

const is_cli = typeof globalThis.open === 'undefined'
const RAW = is_cli ? null : require('random-access-web')

module.exports = identity

function identity (config = {}) {
  const {
    name = 'browser-peer'
  } = config

  // Create corestore for this identity
  const store = is_cli
    ? new Corestore(`./storage-identity-${name}`)
    : new Corestore(RAW(`identity-${name}`))

  // Global state for this identity instance
  const state = {
    events_drive: null,
    events_store: null,
    ds_manager: null,
    pairing_manager: null,
    swarm: null,
    dht: null,
    keypair: null,
    vault_bee: null, // Key-value store for vault data
    vault_audit: null, // Audit log for all vault operations
    store: null, // Store reference for vault operations
    pending_auth: null, // Pending authentication data (for pair mode)
    app_audits: new Map(), // app_id -> auditcore instance
    emitter: make_emitter()
  }

  // vault core keys, for filtering
  const vault_core_keys = new Set()

  // Store config for later use
  const vault_config = config

  // User auth promise, resolves when authentication completes
  let user_resolve, user_reject
  const user = new Promise((resolve, reject) => {
    user_resolve = resolve
    user_reject = reject
  })

  // Create Noise Keypair (Centrally managed via crypto-helpers)
  const user_promise = get_or_create_keypair('vault')

  // Create vault operations (vault API, events, devices, raw data)
  const vault = create_vault_ops(() => state)

  // Return vault object with all identity functions (app-independent only)
  const vault_api = {
    // Config for apps to access
    config: vault_config,
    // User promise (replaces login)
    user,
    authenticate,
    // Vault-only pairing
    create_vault_invite,
    setup_vault_pairing,
    verify_vault_pairing,
    deny_vault_pairing,
    restore_pending_invites,
    cancel_vault_invite,
    // Data structure management
    create_ds_manager,
    // Vault structure initialization (for apps to create vault in seed mode)
    init_vault_structures,
    // Vault operations (delegated to vault-ops module)
    log_event: vault.log_event,
    get_events: vault.get_events,
    get_paired_devices: vault.get_paired_devices,
    get_device_name: vault.get_device_name,
    remove_device: vault.remove_device,
    get_raw_data: vault.get_raw_data,
    register_app: vault.register_app,
    load_app_audit: vault.load_app_audit,
    get_app: vault.get_app,
    list_apps: vault.list_apps,
    vault_put: vault.vault_put,
    vault_get: vault.vault_get,
    vault_del: vault.vault_del,
    // Writer access management
    start_writer_watcher: vault.start_writer_watcher,
    request_writer_access: vault.request_writer_access,
    wait_for_writer_access: vault.wait_for_writer_access,
    log_bootstrap_device: vault.log_bootstrap_device,
    // Complete authentication (called by websys-ui after app join flow)
    complete_authentication,
    // Setters
    set_events_drive,
    set_ds_manager,
    set_pairing_manager,
    set_swarm,
    set_dht,
    get_events_drive,
    get_events_store,
    store,
    get_swarm,
    get_dht,
    get_ds_manager,
    get_vault_bee,
    get_vault_audit,
    get_app_audit,
    // App networking — creates vault-filtered swarm for apps
    network: create_app_network_with_relay,
    // Events
    on_update: handle_update_callback,
    on_vault_ready
  }
  // sessio api is system-only used here but never forwarded to apps
  const session_api = {
    session_get_username,
    session_set_username,
    session_get_auth_mode,
    session_set_auth_mode,
    session_get_joined_app,
    session_set_joined_app,
    session_get_pending_invite,
    session_set_pending_invite,
    session_clear_pending_invite,
    session_clear,
    reset_all_data
  }
  return { ...vault_api, ...session_api }

  /***************************************
INTERNAL FUNCTIONS
***************************************/

  // Authenticate is called by the App's UI. It autheticates the user based on the mode.
  // If user is new, it resolves quickly. If not, it sets up networking to handle pairing.
  async function authenticate (auth_data) {
    const { username, mode, invite_code, on_verification_code } = auth_data
    try {
      if (mode === 'pair' && invite_code) {
        const invite_data = parse_vault_invite(invite_code)
        const mnemonic_data = await user_promise
        state.keypair = mnemonic_data
        const relay_url = await get_default_relay()
        const { swarm: swarm_instance } = await network(store, mnemonic_data, relay_url)
        state.store = store
        state.swarm = swarm_instance
        state.pairing_manager = pairing_manager_constructor(state.swarm)
        await init_vault_structures({ store, vault_bee_key: invite_data.vault_bee_key, vault_audit_key: invite_data.vault_audit_key })
        const result = await state.pairing_manager.join_with_invite({
          invite_code,
          vault_bee: state.vault_bee,
          vault_audit: state.vault_audit,
          on_verification_code_ready: (code) => {
            if (on_verification_code) on_verification_code(code)
          }
        })
        session_clear_pending_invite() // pairing complete, remove stored invite from session
        state.pending_auth = {
          username: result.username || username,
          mode
        }
        state.emitter.emit('vault_ready', state.pending_auth)
        return
      }
      // Seed mode
      const mnemonic_data = await user_promise
      state.keypair = mnemonic_data
      await init_vault_structures({ store, vault_bee_key: null, vault_audit_key: null })
      await store_device_noise_key(mnemonic_data.keypair.publicKey.toString('hex'))
      const relay_url = await get_default_relay()
      const { swarm } = await network(store, mnemonic_data, relay_url)
      // Reconnect to known devices + store new device keys
      const known = await get_other_device_keys(mnemonic_data.keypair.publicKey)
      for (const key of known) swarm.joinPeer(key)
      swarm.on('connection', (socket) => store_device_noise_key(b4a.toString(socket.remotePublicKey, 'hex')))
      state.store = store
      state.swarm = swarm
      if (auth_data.defer_resolve) {
        state.pending_auth = { username, mode }
        state.emitter.emit('vault_ready', state.pending_auth)
        return
      }
      if (user_resolve) user_resolve(make_app_vault(username, mode))
    } catch (err) {
      console.error('[identity] Authentication error:', err)
      if (user_reject) user_reject(err)
    }
  }

  function complete_authentication () {
    if (!state.pending_auth) return
    if (user_resolve) user_resolve(make_app_vault(state.pending_auth.username, state.pending_auth.mode))
    state.pending_auth = null
  }

  /***************************************
VAULT STRUCTURE INITIALIZATION
***************************************/

  async function init_vault_structures (options) {
    if (state.vault_bee) return // already initialized
    const { store: _store, vault_bee_key, vault_audit_key } = options
    state.store = _store
    state.vault_bee = create_autobee({
      store: _store.namespace('vault-bee'),
      bootstrap: vault_bee_key ? b4a.from(vault_bee_key, 'hex') : null
    })
    state.vault_audit = create_auditcore({
      store: _store.namespace('vault-audit'),
      bootstrap: vault_audit_key ? b4a.from(vault_audit_key, 'hex') : null
    })
    await state.vault_bee.ready()
    await state.vault_audit.ready()
    // every core that exists now is a vault core
    for (const [dk_hex] of store.cores) {
      vault_core_keys.add(dk_hex)
    }
    store._is_vault_core = is_vault_core
  }

  /***************************************
  RELAY CONFIG
  ***************************************/

  async function get_default_relay () {
    if (!state.vault_bee) return null
    const entry = await state.vault_bee.get('config/default_relay')
    return entry?.value || null
  }

  async function create_app_network_with_relay () {
    const relay_url = await get_default_relay()
    return create_app_network(store, relay_url)
  }

  /***************************************
DEVICE NOISE KEY MANAGEMENT
***************************************/

  async function get_other_device_keys (own_pubkey) {
    const keys = []
    for await (const entry of state.vault_bee.create_read_stream({ gte: 'devices/noise/', lt: 'devices/noise0' })) {
      const key = b4a.from(entry.key.replace('devices/noise/', ''), 'hex')
      if (!b4a.equals(key, own_pubkey)) keys.push(key)
    }
    return keys
  }

  async function store_device_noise_key (noise_key_hex) {
    const existing = await state.vault_bee.get(`devices/noise/${noise_key_hex}`)
    if (existing) return
    await state.vault_bee.put(`devices/noise/${noise_key_hex}`, { added: Date.now() })
  }

  function is_vault_core (discovery_key_hex) {
    return vault_core_keys.has(discovery_key_hex)
  }

  /***************************************
VAULT PAIRING
***************************************/

  function get_pairing_manager () {
    if (!state.pairing_manager) state.pairing_manager = pairing_manager_constructor(state.swarm)
    return state.pairing_manager
  }

  async function create_vault_invite () {
    if (!state.vault_bee || !state.vault_audit) throw new Error('Vault not initialized')
    const vault_bee_key = b4a.toString(state.vault_bee.key, 'hex')
    const vault_audit_key = b4a.toString(state.vault_audit.key, 'hex')
    const invite_obj = get_pairing_manager().create_invite(state.vault_bee.key)
    const invite_data = { vault_bee_key, vault_audit_key, blind_invite: b4a.toString(invite_obj.invite, 'base64'), vault_only: true }
    const invite_code = b4a.toString(b4a.from(JSON.stringify(invite_data)), 'base64')
    await state.vault_audit.append({ type: 'invite_created', data: {} })
    // Persist invite so we can rejoin the topic after a refresh (removed on verify or deny)
    await state.vault_bee.put('pending_invites/active', {
      invite_code,
      seed_hex: b4a.toString(invite_obj.seed, 'hex'),
      public_key_hex: b4a.toString(invite_obj.publicKey, 'hex'),
      created_at: Date.now()
    })
    return { invite_code, invite: invite_obj, vault_bee_key, vault_audit_key }
  }

  async function setup_vault_pairing (config) {
    const pm = get_pairing_manager()
    await pm.setup_member({ ...config, vault_bee: state.vault_bee, vault_audit: state.vault_audit })
    return pm
  }

  async function verify_vault_pairing (entered_code) {
    if (!state.pairing_manager) throw new Error('Pairing not setup')
    const result = await state.pairing_manager.verify_and_complete_pairing({ entered_verification_code: entered_code, vault_bee: state.vault_bee, vault_audit: state.vault_audit })
    await state.vault_bee.del('pending_invites/active')
    return result
  }

  function deny_vault_pairing () {
    if (!state.pairing_manager) throw new Error('Pairing not setup')
    state.pairing_manager.deny_pairing()
    state.vault_bee.del('pending_invites/active').catch(function (err) { console.warn('[identity] failed to clear pending invite:', err) })
  }

  /***************************************
  RESTORE PENDING INVITES
  reads stored invite from vault, reconstructs the invite_obj
  ***************************************/
  async function restore_pending_invites (config) {
    if (!state.vault_bee) { return null }
    const entry = await state.vault_bee.get('pending_invites/active')
    const stored = entry?.value
    if (!stored) { return null }
    console.log('[identity] restore_pending_invites: found stored invite, created_at:', new Date(stored.created_at).toLocaleString())
    const parsed = parse_vault_invite(stored.invite_code)
    const invite_obj = {
      invite: b4a.from(parsed.blind_invite, 'base64'),
      seed: b4a.from(stored.seed_hex, 'hex'),
      publicKey: b4a.from(stored.public_key_hex, 'hex')
    }
    await setup_vault_pairing({ ...config, invite: invite_obj })
    return stored
  }

  /***************************************
  CANCEL INVITE
  stops the pairing member, clears all
  pending requests and removes the stored invite from vault
  ***************************************/
  async function cancel_vault_invite () {
    if (state.pairing_manager) {
      await state.pairing_manager.close()
      state.pairing_manager = null
    }
    if (state.vault_bee) {
      await state.vault_bee.del('pending_invites/active').catch(function (err) { console.warn('[identity] cancel_vault_invite: failed to delete pending invite:', err) })
    }
  }

  /***************************************
SETTERS
***************************************/

  function set_events_drive (drive, store_instance, setup_callback) {
    state.events_drive = drive
    state.events_store = store_instance
    setup_callback?.()
    state.emitter.emit('update')
  }

  function create_ds_manager () { return datastructure_manager() }
  function set_ds_manager (manager) { state.ds_manager = manager }
  function set_pairing_manager (manager) { state.pairing_manager = manager }
  function set_swarm (swarm_instance) { state.swarm = swarm_instance }
  function set_dht (dht_instance) { state.dht = dht_instance }

  /***************************************
GETTERS
***************************************/

  function get_events_drive () { return state.events_drive }
  function get_events_store () { return state.events_store }
  function get_swarm () { return state.swarm }
  function get_dht () { return state.dht }
  function get_ds_manager () { return state.ds_manager }
  function get_vault_bee () { return state.vault_bee }
  function get_vault_audit () { return state.vault_audit }
  function get_app_audit (app_id) { return state.app_audits.get(app_id) || null }

  /***************************************
EVENT EMITTER
***************************************/

  function handle_update_callback (cb) { return state.emitter.on('update', cb) }
  function on_vault_ready (cb) { return state.emitter.on('vault_ready', cb) }

  /***************************************
  SESSION MANAGEMENT, used for times when localstorage usage is mandatory
  ***************************************/
  function session_get_username () { return localStorage.getItem('username') }
  function session_set_username (v) { localStorage.setItem('username', v) }
  function session_get_auth_mode () { return localStorage.getItem('auth_mode') || 'seed' }
  function session_set_auth_mode (v) { localStorage.setItem('auth_mode', v) }
  function session_get_joined_app () { return localStorage.getItem('joined_app') || '' }
  function session_set_joined_app (v) { localStorage.setItem('joined_app', v) }
  function session_get_pending_invite () { return localStorage.getItem('pending_invite_code') || null }
  function session_set_pending_invite (v) { localStorage.setItem('pending_invite_code', v) }
  function session_clear_pending_invite () { localStorage.removeItem('pending_invite_code') }
  function session_clear () { localStorage.clear() }
  // Apps get vault_api only.. no session methods, no localstorage access
  function make_app_vault (username, mode) {
    return { ...vault_api, username, mode, authenticated: true }
  }
  async function reset_all_data () {
    localStorage.clear()
    if (is_cli) return reset_cli_data()
    return reset_browser_data()
  }

  /* We already have a localstorage shim for CLi environment,
  as cli also uses the same vault apis, in its environment, it will use helpers/local-storage-shim.
  so we don't have to do much here on the vault code. */
  function reset_cli_data () {
    localStorage.reset_all()
  }

  /***************************************
  Reset Browser IndexedDB
  ***************************************/

  async function reset_browser_data () {
    const databases = await indexedDB.databases()
    for (const db of databases) {
      if (db.name && (db.name.includes('blogs-') || db.name.includes('random-access-web') || db.name.includes('identity-'))) {
        indexedDB.deleteDatabase(db.name)
      }
    }
    if (typeof requestFileSystem === 'undefined' && typeof webkitRequestFileSystem === 'undefined') return
    const rfs = typeof requestFileSystem !== 'undefined' ? requestFileSystem : webkitRequestFileSystem
    const persistent = typeof self !== 'undefined' ? self.PERSISTENT : 1
    await new Promise(init_cleanup)
    function init_cleanup (resolve, reject) {
      rfs(persistent, 1024 * 1024, (fs) => read_root(fs, resolve, reject), reject)
    }
    function read_root (fs, resolve, reject) {
      fs.root.createReader().readEntries((entries) => process_entries(entries, resolve), reject)
    }
    function process_entries (entries, resolve) {
      if (!entries.length) return resolve()
      let completed = 0
      const total = entries.length
      function on_done () {
        if (++completed === total) resolve()
      }
      entries.forEach(remove_entry)
      function remove_entry (entry) {
        if (entry.isFile) entry.remove(on_done, on_done)
        else entry.removeRecursively(on_done, on_done)
      }
    }
  }
}
